// @neondatabase/neon-js/dist/index.mjs
import { createInternalNeonAuth } from "@neondatabase/auth";
import { BetterAuthVanillaAdapter, SupabaseAuthAdapter } from "@neondatabase/auth/vanilla/adapters";
import "@neondatabase/auth/react/adapters";
import { AuthRequiredError, NeonPostgrestClient, X_NEON_CLIENT_INFO_HEADER, fetchWithToken, fetchWithToken as fetchWithToken$1, getClientInfo } from "@neondatabase/postgrest-js";

//#region src/client/neon-client.ts
/**
* Neon client with integrated authentication
*
* Extends NeonPostgrestClient with Neon Auth integration.
* For auth-free clients, use @neondatabase/postgrest-js instead.
*/
var NeonClient = class extends NeonPostgrestClient {
	auth;
	constructor({ dataApiUrl, options, authClient }) {
		super({
			dataApiUrl,
			options
		});
		this.auth = authClient.adapter;
	}
};

//#endregion
//#region package.json
var name = "@neondatabase/neon-js";
var version = "0.6.2-beta";

//#endregion
//#region src/utils/client-info.ts
function buildNeonJsClientInfo() {
	const info = getClientInfo(name, version);
	return JSON.stringify(info);
}

//#endregion
//#region src/client/client-factory.ts
function createClient(config) {
	const { auth: authConfig, dataApi: dataApiConfig } = config;
	const clientInfoHeader = buildNeonJsClientInfo();
	const auth = createInternalNeonAuth(authConfig.url, {
		adapter: authConfig.adapter,
		allowAnonymous: authConfig.allowAnonymous ?? false,
		fetchOptions: { headers: { [X_NEON_CLIENT_INFO_HEADER]: clientInfoHeader } }
	});
	const getAccessToken = async () => {
		return auth.getJWTToken();
	};
	const authFetch = fetchWithToken$1(getAccessToken, dataApiConfig.options?.global?.fetch);
	return new NeonClient({
		dataApiUrl: dataApiConfig.url,
		authClient: auth,
		options: {
			...dataApiConfig.options,
			global: {
				...dataApiConfig.options?.global,
				fetch: authFetch,
				headers: {
					...dataApiConfig.options?.global?.headers,
					[X_NEON_CLIENT_INFO_HEADER]: clientInfoHeader
				}
			}
		}
	});
}

//#endregion
export { AuthRequiredError, BetterAuthVanillaAdapter, SupabaseAuthAdapter, createClient, fetchWithToken };
