// @neondatabase/neon-js/dist/cli/index.mjs
#!/usr/bin/env node
import meow from "meow";
import { ident, literal } from "pg-format";
import pg from "pg";
import { parse } from "postgres-array";
import prettier from "prettier";
import { mkdirSync, writeFileSync } from "node:fs";
import path from "node:path";

//#region src/cli/utils/parse-duration.ts
function parseDuration(duration) {
	const match = duration.match(/^(\d+)([hms])?$/);
	if (!match) throw new Error(`Invalid duration format: ${duration}. Use format like "30s", "1m", "90s"`);
	const value = Number.parseInt(match[1], 10);
	const unit = match[2] || "s";
	switch (unit) {
		case "s": return value * 1e3;
		case "m": return value * 60 * 1e3;
		case "h": return value * 60 * 60 * 1e3;
		default: throw new Error(`Unknown duration unit: ${unit}`);
	}
}

//#endregion
//#region src/vendor/postgres-meta/lib/constants.ts
const DEFAULT_SYSTEM_SCHEMAS = [
	"information_schema",
	"pg_catalog",
	"pg_toast"
];

//#endregion
//#region src/vendor/postgres-meta/lib/helpers.ts
const coalesceRowsToArray = (source, filter) => {
	return `
COALESCE(
  (
    SELECT
      array_agg(row_to_json(${source})) FILTER (WHERE ${filter})
    FROM
      ${source}
  ),
  '{}'
) AS ${source}`;
};
const filterByList = (include, exclude, defaultExclude) => {
	if (defaultExclude) exclude = defaultExclude.concat(exclude ?? []);
	if (include?.length) return `IN (${include.map(literal).join(",")})`;
	else if (exclude?.length) return `NOT IN (${exclude.map(literal).join(",")})`;
	return "";
};
const filterByValue = (ids) => {
	if (ids?.length) return `IN (${ids.map(literal).join(",")})`;
	return "";
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/column_privileges.sql.ts
const COLUMN_PRIVILEGES_SQL = (props) => `
-- Lists each column's privileges in the form of:
--
-- [
--   {
--     "column_id": "12345.1",
--     "relation_schema": "public",
--     "relation_name": "mytable",
--     "column_name": "mycolumn",
--     "privileges": [
--       {
--         "grantor": "postgres",
--         "grantee": "myrole",
--         "privilege_type": "SELECT",
--         "is_grantable": false
--       },
--       ...
--     ]
--   },
--   ...
-- ]
--
-- Modified from information_schema.column_privileges. We try to be as close as
-- possible to the view definition, obtained from:
--
-- select pg_get_viewdef('information_schema.column_privileges');
--
-- The main differences are:
-- - we include column privileges for materialized views
--   (reason for exclusion in information_schema.column_privileges:
--    https://www.postgresql.org/message-id/9136.1502740844%40sss.pgh.pa.us)
-- - we query a.attrelid and a.attnum to generate \`column_id\`
-- - \`table_catalog\` is omitted
-- - table_schema -> relation_schema, table_name -> relation_name
--
-- Column privileges are intertwined with table privileges in that table
-- privileges override column privileges. E.g. if we do:
--
-- grant all on mytable to myrole;
--
-- Then \`myrole\` is granted privileges for ALL columns. Likewise, if we do:
--
-- grant all (id) on mytable to myrole;
-- revoke all on mytable from myrole;
--
-- Then the grant on the \`id\` column is revoked.
--
-- This is unlike how grants for schemas and tables interact, where you need
-- privileges for BOTH the schema the table is in AND the table itself in order
-- to access the table.

select (x.attrelid || '.' || x.attnum) as column_id,
       nc.nspname as relation_schema,
       x.relname as relation_name,
       x.attname as column_name,
       coalesce(
         jsonb_agg(
           jsonb_build_object(
             'grantor', u_grantor.rolname,
             'grantee', grantee.rolname,
             'privilege_type', x.prtype,
             'is_grantable', x.grantable
           )
         ),
         '[]'
       ) as privileges
from
  (select pr_c.grantor,
          pr_c.grantee,
          a.attrelid,
          a.attnum,
          a.attname,
          pr_c.relname,
          pr_c.relnamespace,
          pr_c.prtype,
          pr_c.grantable,
          pr_c.relowner
   from
     (select pg_class.oid,
             pg_class.relname,
             pg_class.relnamespace,
             pg_class.relowner,
             (aclexplode(coalesce(pg_class.relacl, acldefault('r', pg_class.relowner)))).grantor as grantor,
             (aclexplode(coalesce(pg_class.relacl, acldefault('r', pg_class.relowner)))).grantee as grantee,
             (aclexplode(coalesce(pg_class.relacl, acldefault('r', pg_class.relowner)))).privilege_type as privilege_type,
             (aclexplode(coalesce(pg_class.relacl, acldefault('r', pg_class.relowner)))).is_grantable as is_grantable
      from pg_class
      where (pg_class.relkind = any (array['r',
                                           'v',
                                           'm',
                                           'f',
                                           'p'])) ) pr_c(oid, relname, relnamespace, relowner, grantor, grantee, prtype, grantable),
                                                    pg_attribute a
   where ((a.attrelid = pr_c.oid)
          and (a.attnum > 0)
          and (not a.attisdropped))
   union select pr_a.grantor,
                pr_a.grantee,
                pr_a.attrelid,
                pr_a.attnum,
                pr_a.attname,
                c.relname,
                c.relnamespace,
                pr_a.prtype,
                pr_a.grantable,
                c.relowner
   from
     (select a.attrelid,
             a.attnum,
             a.attname,
             (aclexplode(coalesce(a.attacl, acldefault('c', cc.relowner)))).grantor as grantor,
             (aclexplode(coalesce(a.attacl, acldefault('c', cc.relowner)))).grantee as grantee,
             (aclexplode(coalesce(a.attacl, acldefault('c', cc.relowner)))).privilege_type as privilege_type,
             (aclexplode(coalesce(a.attacl, acldefault('c', cc.relowner)))).is_grantable as is_grantable
      from (pg_attribute a
            join pg_class cc on ((a.attrelid = cc.oid)))
      where ((a.attnum > 0)
             and (not a.attisdropped))) pr_a(attrelid, attnum, attname, grantor, grantee, prtype, grantable),
                                        pg_class c
   where ((pr_a.attrelid = c.oid)
          and (c.relkind = any (ARRAY['r',
                                      'v',
                                      'm',
                                      'f',
                                      'p'])))) x,
     pg_namespace nc,
     pg_authid u_grantor,
  (select pg_authid.oid,
          pg_authid.rolname
   from pg_authid
   union all select (0)::oid as oid,
                    'PUBLIC') grantee(oid, rolname)
where ((x.relnamespace = nc.oid)
       ${props.schemaFilter ? `and nc.nspname ${props.schemaFilter}` : ""}
       ${props.columnIdsFilter ? `and (x.attrelid || '.' || x.attnum) ${props.columnIdsFilter}` : ""}
       and (x.grantee = grantee.oid)
       and (x.grantor = u_grantor.oid)
       and (x.prtype = any (ARRAY['INSERT',
                                  'SELECT',
                                  'UPDATE',
                                  'REFERENCES']))
       and (pg_has_role(u_grantor.oid, 'USAGE')
            or pg_has_role(grantee.oid, 'USAGE')
            or (grantee.rolname = 'PUBLIC')))
group by column_id,
         nc.nspname,
         x.relname,
         x.attname
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaColumnPrivileges.ts
var PostgresMetaColumnPrivileges = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset } = {}) {
		const sql = COLUMN_PRIVILEGES_SQL({
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async grant(grants) {
		let sql = `
do $$
declare
  col record;
begin
${grants.map(({ privilege_type, column_id, grantee, is_grantable }) => {
			const [relationId, columnNumber] = column_id.split(".");
			return `
select *
from pg_attribute a
where a.attrelid = ${literal(relationId)}
  and a.attnum = ${literal(columnNumber)}
into col;
execute format(
  'grant ${privilege_type} (%I) on %s to ${grantee.toLowerCase() === "public" ? "public" : ident(grantee)} ${is_grantable ? "with grant option" : ""}',
  col.attname,
  col.attrelid::regclass
);`;
		}).join("\n")}
end $$;
`;
		const { data, error } = await this.query(sql);
		if (error) return {
			data,
			error
		};
		sql = COLUMN_PRIVILEGES_SQL({ columnIdsFilter: filterByValue([...new Set(grants.map(({ column_id }) => column_id))]) });
		return await this.query(sql);
	}
	async revoke(revokes) {
		let sql = `
do $$
declare
  col record;
begin
${revokes.map(({ privilege_type, column_id, grantee }) => {
			const [relationId, columnNumber] = column_id.split(".");
			return `
select *
from pg_attribute a
where a.attrelid = ${literal(relationId)}
  and a.attnum = ${literal(columnNumber)}
into col;
execute format(
  'revoke ${privilege_type} (%I) on %s from ${grantee.toLowerCase() === "public" ? "public" : ident(grantee)}',
  col.attname,
  col.attrelid::regclass
);`;
		}).join("\n")}
end $$;
`;
		const { data, error } = await this.query(sql);
		if (error) return {
			data,
			error
		};
		sql = COLUMN_PRIVILEGES_SQL({ columnIdsFilter: filterByValue([...new Set(revokes.map(({ column_id }) => column_id))]) });
		return await this.query(sql);
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/table.sql.ts
const TABLES_SQL = (props) => `
SELECT
  c.oid :: int8 AS id,
  nc.nspname AS schema,
  c.relname AS name,
  c.relrowsecurity AS rls_enabled,
  c.relforcerowsecurity AS rls_forced,
  CASE
    WHEN c.relreplident = 'd' THEN 'DEFAULT'
    WHEN c.relreplident = 'i' THEN 'INDEX'
    WHEN c.relreplident = 'f' THEN 'FULL'
    ELSE 'NOTHING'
  END AS replica_identity,
  pg_total_relation_size(format('%I.%I', nc.nspname, c.relname)) :: int8 AS bytes,
  pg_size_pretty(
    pg_total_relation_size(format('%I.%I', nc.nspname, c.relname))
  ) AS size,
  pg_stat_get_live_tuples(c.oid) AS live_rows_estimate,
  pg_stat_get_dead_tuples(c.oid) AS dead_rows_estimate,
  obj_description(c.oid) AS comment,
  coalesce(pk.primary_keys, '[]') as primary_keys,
  coalesce(
    jsonb_agg(relationships) filter (where relationships is not null),
    '[]'
  ) as relationships
FROM
  pg_namespace nc
  JOIN pg_class c ON nc.oid = c.relnamespace
  left join (
    select
      c.oid::int8 as table_id,
      jsonb_agg(
        jsonb_build_object(
          'table_id', c.oid::int8,
          'schema', n.nspname,
          'table_name', c.relname,
          'name', a.attname
        )
        order by array_position(i.indkey, a.attnum)
      ) as primary_keys
    from
      pg_index i
      join pg_class c on i.indrelid = c.oid
      join pg_namespace n on c.relnamespace = n.oid
      join pg_attribute a on a.attrelid = c.oid and a.attnum = any(i.indkey)
    where
      ${props.schemaFilter ? `n.nspname ${props.schemaFilter} AND` : ""}
      ${props.tableIdentifierFilter ? `n.nspname || '.' || c.relname ${props.tableIdentifierFilter} AND` : ""}
      i.indisprimary
    group by c.oid
  ) as pk
  on pk.table_id = c.oid
  left join (
    select
      c.oid :: int8 as id,
      c.conname as constraint_name,
      nsa.nspname as source_schema,
      csa.relname as source_table_name,
      sa.attname as source_column_name,
      nta.nspname as target_table_schema,
      cta.relname as target_table_name,
      ta.attname as target_column_name
    from
      pg_constraint c
    join (
      pg_attribute sa
      join pg_class csa on sa.attrelid = csa.oid
      join pg_namespace nsa on csa.relnamespace = nsa.oid
    ) on sa.attrelid = c.conrelid and sa.attnum = any (c.conkey)
    join (
      pg_attribute ta
      join pg_class cta on ta.attrelid = cta.oid
      join pg_namespace nta on cta.relnamespace = nta.oid
    ) on ta.attrelid = c.confrelid and ta.attnum = any (c.confkey)
    where
      ${props.schemaFilter ? `nsa.nspname ${props.schemaFilter} OR nta.nspname ${props.schemaFilter} AND` : ""}
      ${props.tableIdentifierFilter ? `(nsa.nspname || '.' || csa.relname) ${props.tableIdentifierFilter} OR (nta.nspname || '.' || cta.relname) ${props.tableIdentifierFilter} AND` : ""}
      c.contype = 'f'
  ) as relationships
  on (relationships.source_schema = nc.nspname and relationships.source_table_name = c.relname)
  or (relationships.target_table_schema = nc.nspname and relationships.target_table_name = c.relname)
WHERE
  ${props.schemaFilter ? `nc.nspname ${props.schemaFilter} AND` : ""}
  ${props.idsFilter ? `c.oid ${props.idsFilter} AND` : ""}
  ${props.tableIdentifierFilter ? `nc.nspname || '.' || c.relname ${props.tableIdentifierFilter} AND` : ""}
  c.relkind IN ('r', 'p')
  AND NOT pg_is_other_temp_schema(nc.oid)
  AND (
    pg_has_role(c.relowner, 'USAGE')
    OR has_table_privilege(
      c.oid,
      'SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, TRIGGER'
    )
    OR has_any_column_privilege(c.oid, 'SELECT, INSERT, UPDATE, REFERENCES')
  )
group by
  c.oid,
  c.relname,
  c.relrowsecurity,
  c.relforcerowsecurity,
  c.relreplident,
  nc.nspname,
  pk.primary_keys
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/sql/columns.sql.ts
const COLUMNS_SQL = (props) => `
-- Adapted from information_schema.columns

SELECT
  c.oid :: int8 AS table_id,
  nc.nspname AS schema,
  c.relname AS table,
  (c.oid || '.' || a.attnum) AS id,
  a.attnum AS ordinal_position,
  a.attname AS name,
  CASE
    WHEN a.atthasdef THEN pg_get_expr(ad.adbin, ad.adrelid)
    ELSE NULL
  END AS default_value,
  CASE
    WHEN t.typtype = 'd' THEN CASE
      WHEN bt.typelem <> 0 :: oid
      AND bt.typlen = -1 THEN 'ARRAY'
      WHEN nbt.nspname = 'pg_catalog' THEN format_type(t.typbasetype, NULL)
      ELSE 'USER-DEFINED'
    END
    ELSE CASE
      WHEN t.typelem <> 0 :: oid
      AND t.typlen = -1 THEN 'ARRAY'
      WHEN nt.nspname = 'pg_catalog' THEN format_type(a.atttypid, NULL)
      ELSE 'USER-DEFINED'
    END
  END AS data_type,
  COALESCE(bt.typname, t.typname) AS format,
  a.attidentity IN ('a', 'd') AS is_identity,
  CASE
    a.attidentity
    WHEN 'a' THEN 'ALWAYS'
    WHEN 'd' THEN 'BY DEFAULT'
    ELSE NULL
  END AS identity_generation,
  a.attgenerated IN ('s') AS is_generated,
  NOT (
    a.attnotnull
    OR t.typtype = 'd' AND t.typnotnull
  ) AS is_nullable,
  (
    c.relkind IN ('r', 'p')
    OR c.relkind IN ('v', 'f') AND pg_column_is_updatable(c.oid, a.attnum, FALSE)
  ) AS is_updatable,
  uniques.table_id IS NOT NULL AS is_unique,
  check_constraints.definition AS "check",
  array_to_json(
    array(
      SELECT
        enumlabel
      FROM
        pg_catalog.pg_enum enums
      WHERE
        enums.enumtypid = coalesce(bt.oid, t.oid)
        OR enums.enumtypid = coalesce(bt.typelem, t.typelem)
      ORDER BY
        enums.enumsortorder
    )
  ) AS enums,
  col_description(c.oid, a.attnum) AS comment
FROM
  pg_attribute a
  LEFT JOIN pg_attrdef ad ON a.attrelid = ad.adrelid
  AND a.attnum = ad.adnum
  JOIN (
    pg_class c
    JOIN pg_namespace nc ON c.relnamespace = nc.oid
  ) ON a.attrelid = c.oid
  JOIN (
    pg_type t
    JOIN pg_namespace nt ON t.typnamespace = nt.oid
  ) ON a.atttypid = t.oid
  LEFT JOIN (
    pg_type bt
    JOIN pg_namespace nbt ON bt.typnamespace = nbt.oid
  ) ON t.typtype = 'd'
  AND t.typbasetype = bt.oid
  LEFT JOIN (
    SELECT DISTINCT ON (table_id, ordinal_position)
      conrelid AS table_id,
      conkey[1] AS ordinal_position
    FROM pg_catalog.pg_constraint
    WHERE contype = 'u' AND cardinality(conkey) = 1
  ) AS uniques ON uniques.table_id = c.oid AND uniques.ordinal_position = a.attnum
  LEFT JOIN (
    -- We only select the first column check
    SELECT DISTINCT ON (table_id, ordinal_position)
      conrelid AS table_id,
      conkey[1] AS ordinal_position,
      substring(
        pg_get_constraintdef(pg_constraint.oid, true),
        8,
        length(pg_get_constraintdef(pg_constraint.oid, true)) - 8
      ) AS "definition"
    FROM pg_constraint
    WHERE contype = 'c' AND cardinality(conkey) = 1
    ORDER BY table_id, ordinal_position, oid asc
  ) AS check_constraints ON check_constraints.table_id = c.oid AND check_constraints.ordinal_position = a.attnum
WHERE
  ${props.schemaFilter ? `nc.nspname ${props.schemaFilter} AND` : ""}
  ${props.idsFilter ? `(c.oid || '.' || a.attnum) ${props.idsFilter} AND` : ""}
  ${props.columnNameFilter ? `(c.relname || '.' || a.attname) ${props.columnNameFilter} AND` : ""}
  ${props.tableIdFilter ? `c.oid ${props.tableIdFilter} AND` : ""}
  ${props.tableIdentifierFilter ? `nc.nspname || '.' || c.relname ${props.tableIdentifierFilter} AND` : ""}
  NOT pg_is_other_temp_schema(nc.oid)
  AND a.attnum > 0
  AND NOT a.attisdropped
  AND (c.relkind IN ('r', 'v', 'm', 'f', 'p'))
  AND (
    pg_has_role(c.relowner, 'USAGE')
    OR has_column_privilege(
      c.oid,
      a.attnum,
      'SELECT, INSERT, UPDATE, REFERENCES'
    )
  )
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaTables.ts
var PostgresMetaTables = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset, includeColumns = true } = {}) {
		const sql = generateEnrichedTablesSql({
			includeColumns,
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name, schema = "public" }) {
		const schemaFilter = schema ? filterByList([schema], []) : void 0;
		if (id) {
			const sql = generateEnrichedTablesSql({
				schemaFilter,
				includeColumns: true,
				idsFilter: filterByValue([id])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a table with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name) {
			const sql = generateEnrichedTablesSql({
				schemaFilter,
				includeColumns: true,
				tableIdentifierFilter: filterByValue([`${schema}.${name}`])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a table named ${name} in schema ${schema}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on table retrieve" }
		};
	}
	async create({ name, schema = "public", comment }) {
		const sql = `BEGIN; ${`CREATE TABLE ${ident(schema)}.${ident(name)} ();`} ${comment === void 0 ? "" : `COMMENT ON TABLE ${ident(schema)}.${ident(name)} IS ${literal(comment)};`} COMMIT;`;
		const { error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return await this.retrieve({
			name,
			schema
		});
	}
	async update(id, { name, schema, rls_enabled, rls_forced, replica_identity, replica_identity_index, primary_keys, comment }) {
		const { data: old, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const alter = `ALTER TABLE ${ident(old.schema)}.${ident(old.name)}`;
		const schemaSql = schema === void 0 ? "" : `${alter} SET SCHEMA ${ident(schema)};`;
		let nameSql = "";
		if (name !== void 0 && name !== old.name) nameSql = `ALTER TABLE ${ident(schema === void 0 ? old.schema : schema)}.${ident(old.name)} RENAME TO ${ident(name)};`;
		let enableRls = "";
		if (rls_enabled !== void 0) {
			const enable = `${alter} ENABLE ROW LEVEL SECURITY;`;
			const disable = `${alter} DISABLE ROW LEVEL SECURITY;`;
			enableRls = rls_enabled ? enable : disable;
		}
		let forceRls = "";
		if (rls_forced !== void 0) {
			const enable = `${alter} FORCE ROW LEVEL SECURITY;`;
			const disable = `${alter} NO FORCE ROW LEVEL SECURITY;`;
			forceRls = rls_forced ? enable : disable;
		}
		let replicaSql = "";
		if (replica_identity === void 0) {} else if (replica_identity === "INDEX") replicaSql = `${alter} REPLICA IDENTITY USING INDEX ${replica_identity_index};`;
		else replicaSql = `${alter} REPLICA IDENTITY ${replica_identity};`;
		let primaryKeysSql = "";
		if (primary_keys === void 0) {} else {
			if (old.primary_keys.length !== 0) primaryKeysSql += `
DO $$
DECLARE
  r record;
BEGIN
  SELECT conname
    INTO r
    FROM pg_constraint
    WHERE contype = 'p' AND conrelid = ${literal(id)};
  EXECUTE ${literal(`${alter} DROP CONSTRAINT `)} || quote_ident(r.conname);
END
$$;
`;
			if (primary_keys.length === 0) {} else primaryKeysSql += `${alter} ADD PRIMARY KEY (${primary_keys.map((x) => ident(x.name)).join(",")});`;
		}
		const commentSql = comment === void 0 ? "" : `COMMENT ON TABLE ${ident(old.schema)}.${ident(old.name)} IS ${literal(comment)};`;
		const sql = `
BEGIN;
  ${enableRls}
  ${forceRls}
  ${replicaSql}
  ${primaryKeysSql}
  ${commentSql}
  ${schemaSql}
  ${nameSql}
COMMIT;`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return await this.retrieve({ id });
	}
	async remove(id, { cascade = false } = {}) {
		const { data: table, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const sql = `DROP TABLE ${ident(table.schema)}.${ident(table.name)} ${cascade ? "CASCADE" : "RESTRICT"};`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return {
			data: table,
			error: null
		};
	}
};
const generateEnrichedTablesSql = ({ includeColumns, schemaFilter, tableIdentifierFilter, idsFilter, limit, offset }) => `
with tables as (${TABLES_SQL({
	schemaFilter,
	tableIdentifierFilter,
	idsFilter,
	limit,
	offset
})})
  ${includeColumns ? `, columns as (${COLUMNS_SQL({
	schemaFilter,
	tableIdFilter: idsFilter,
	tableIdentifierFilter
})})` : ""}
select
  *
  ${includeColumns ? `, ${coalesceRowsToArray("columns", "columns.table_id = tables.id")}` : ""}
from tables`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaColumns.ts
var PostgresMetaColumns = class {
	query;
	metaTables;
	constructor(query) {
		this.query = query;
		this.metaTables = new PostgresMetaTables(query);
	}
	async list({ tableId, includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset } = {}) {
		const sql = COLUMNS_SQL({
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			tableIdFilter: tableId ? filterByValue([`${tableId}`]) : void 0,
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name, table, schema = "public" }) {
		const schemaFilter = schema ? filterByList([schema], []) : void 0;
		if (id) {
			const regexp = /^(\d+)\.(\d+)$/;
			if (!regexp.test(id)) return {
				data: null,
				error: { message: "Invalid format for column ID" }
			};
			const [tableId, ordinalPos] = id.match(regexp).slice(1).map(Number);
			const sql = COLUMNS_SQL({ idsFilter: filterByValue([`${tableId}.${ordinalPos}`]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a column with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name && table) {
			const sql = `${COLUMNS_SQL({
				schemaFilter,
				columnNameFilter: filterByValue([`${table}.${name}`])
			})};`;
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a column named ${name} in table ${schema}.${table}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on column retrieve" }
		};
	}
	async create({ table_id, name, type, default_value, default_value_format = "literal", is_identity = false, identity_generation = "BY DEFAULT", is_nullable, is_primary_key = false, is_unique = false, comment, check }) {
		const { data, error } = await this.metaTables.retrieve({ id: table_id });
		if (error) return {
			data: null,
			error
		};
		const { name: table, schema } = data;
		let defaultValueClause = "";
		if (is_identity) {
			if (default_value !== void 0) return {
				data: null,
				error: { message: "Columns cannot both be identity and have a default value" }
			};
			defaultValueClause = `GENERATED ${identity_generation} AS IDENTITY`;
		} else if (default_value === void 0) {} else if (default_value_format === "expression") defaultValueClause = `DEFAULT ${default_value}`;
		else defaultValueClause = `DEFAULT ${literal(default_value)}`;
		let isNullableClause = "";
		if (is_nullable !== void 0) isNullableClause = is_nullable ? "NULL" : "NOT NULL";
		const isPrimaryKeyClause = is_primary_key ? "PRIMARY KEY" : "";
		const isUniqueClause = is_unique ? "UNIQUE" : "";
		const checkSql = check === void 0 ? "" : `CHECK (${check})`;
		const commentSql = comment === void 0 ? "" : `COMMENT ON COLUMN ${ident(schema)}.${ident(table)}.${ident(name)} IS ${literal(comment)}`;
		const sql = `
BEGIN;
  ALTER TABLE ${ident(schema)}.${ident(table)} ADD COLUMN ${ident(name)} ${typeIdent(type)}
    ${defaultValueClause}
    ${isNullableClause}
    ${isPrimaryKeyClause}
    ${isUniqueClause}
    ${checkSql};
  ${commentSql};
COMMIT;`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return await this.retrieve({
			name,
			table,
			schema
		});
	}
	async update(id, { name, type, drop_default = false, default_value, default_value_format = "literal", is_identity, identity_generation = "BY DEFAULT", is_nullable, is_unique, comment, check }) {
		const { data: old, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const nameSql = name === void 0 || name === old.name ? "" : `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} RENAME COLUMN ${ident(old.name)} TO ${ident(name)};`;
		const typeSql = type === void 0 ? "" : `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ALTER COLUMN ${ident(old.name)} SET DATA TYPE ${typeIdent(type)} USING ${ident(old.name)}::${typeIdent(type)};`;
		let defaultValueSql;
		if (drop_default) defaultValueSql = `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ALTER COLUMN ${ident(old.name)} DROP DEFAULT;`;
		else if (default_value === void 0) defaultValueSql = "";
		else {
			const defaultValue = default_value_format === "expression" ? default_value : literal(default_value);
			defaultValueSql = `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ALTER COLUMN ${ident(old.name)} SET DEFAULT ${defaultValue};`;
		}
		let identitySql = `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ALTER COLUMN ${ident(old.name)}`;
		if (is_identity === false) identitySql += " DROP IDENTITY IF EXISTS;";
		else if (old.is_identity === true) if (identity_generation === void 0) identitySql = "";
		else identitySql += ` SET GENERATED ${identity_generation};`;
		else if (is_identity === void 0) identitySql = "";
		else identitySql += ` ADD GENERATED ${identity_generation} AS IDENTITY;`;
		let isNullableSql;
		if (is_nullable === void 0) isNullableSql = "";
		else isNullableSql = is_nullable ? `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ALTER COLUMN ${ident(old.name)} DROP NOT NULL;` : `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ALTER COLUMN ${ident(old.name)} SET NOT NULL;`;
		let isUniqueSql = "";
		if (old.is_unique === true && is_unique === false) isUniqueSql = `
DO $$
DECLARE
  r record;
BEGIN
  FOR r IN
    SELECT conname FROM pg_constraint WHERE
      contype = 'u'
      AND cardinality(conkey) = 1
      AND conrelid = ${literal(old.table_id)}
      AND conkey[1] = ${literal(old.ordinal_position)}
  LOOP
    EXECUTE ${literal(`ALTER TABLE ${ident(old.schema)}.${ident(old.table)} DROP CONSTRAINT `)} || quote_ident(r.conname);
  END LOOP;
END
$$;
`;
		else if (old.is_unique === false && is_unique === true) isUniqueSql = `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ADD UNIQUE (${ident(old.name)});`;
		const commentSql = comment === void 0 ? "" : `COMMENT ON COLUMN ${ident(old.schema)}.${ident(old.table)}.${ident(old.name)} IS ${literal(comment)};`;
		const checkSql = check === void 0 ? "" : `
DO $$
DECLARE
  v_conname name;
  v_conkey int2[];
BEGIN
  SELECT conname into v_conname FROM pg_constraint WHERE
    contype = 'c'
    AND cardinality(conkey) = 1
    AND conrelid = ${literal(old.table_id)}
    AND conkey[1] = ${literal(old.ordinal_position)}
    ORDER BY oid asc
    LIMIT 1;

  IF v_conname IS NOT NULL THEN
    EXECUTE format('ALTER TABLE ${ident(old.schema)}.${ident(old.table)} DROP CONSTRAINT %s', v_conname);
  END IF;

  ${check !== null ? `
  ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ADD CONSTRAINT ${ident(`${old.table}_${old.name}_check`)} CHECK (${check});

  SELECT conkey into v_conkey FROM pg_constraint WHERE conname = ${literal(`${old.table}_${old.name}_check`)};

  ASSERT v_conkey IS NOT NULL, 'error creating column constraint: check condition must refer to this column';
  ASSERT cardinality(v_conkey) = 1, 'error creating column constraint: check condition cannot refer to multiple columns';
  ASSERT v_conkey[1] = ${literal(old.ordinal_position)}, 'error creating column constraint: check condition cannot refer to other columns';
  ` : ""}
END
$$;
`;
		const sql = `
BEGIN;
  ${isNullableSql}
  ${typeSql}
  ${defaultValueSql}
  ${identitySql}
  ${isUniqueSql}
  ${commentSql}
  ${checkSql}
  ${nameSql}
COMMIT;`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return await this.retrieve({ id });
	}
	async remove(id, { cascade = false } = {}) {
		const { data: column, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const sql = `ALTER TABLE ${ident(column.schema)}.${ident(column.table)} DROP COLUMN ${ident(column.name)} ${cascade ? "CASCADE" : "RESTRICT"};`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return {
			data: column,
			error: null
		};
	}
};
const typeIdent = (type) => {
	return type.endsWith("[]") ? `${ident(type.slice(0, -2))}[]` : type.includes(".") ? type : ident(type);
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/config.sql.ts
const CONFIG_SQL = (props) => `
SELECT
  name,
  setting,
  category,
  TRIM(split_part(category, '/', 1)) AS group,
  TRIM(split_part(category, '/', 2)) AS subgroup,
  unit,
  short_desc,
  extra_desc,
  context,
  vartype,
  source,
  min_val,
  max_val,
  enumvals,
  boot_val,
  reset_val,
  sourcefile,
  sourceline,
  pending_restart
FROM
  pg_settings
ORDER BY
  category,
  name
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaConfig.ts
var PostgresMetaConfig = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ limit, offset } = {}) {
		const sql = CONFIG_SQL({
			limit,
			offset
		});
		return await this.query(sql);
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/extensions.sql.ts
const EXTENSIONS_SQL = (props) => `
SELECT
  e.name,
  n.nspname AS schema,
  e.default_version,
  x.extversion AS installed_version,
  e.comment
FROM
  pg_available_extensions() e(name, default_version, comment)
  LEFT JOIN pg_extension x ON e.name = x.extname
  LEFT JOIN pg_namespace n ON x.extnamespace = n.oid
WHERE
  true
  ${props.nameFilter ? `AND e.name ${props.nameFilter}` : ""}
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaExtensions.ts
var PostgresMetaExtensions = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ limit, offset } = {}) {
		const sql = EXTENSIONS_SQL({
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ name }) {
		const sql = EXTENSIONS_SQL({ nameFilter: filterByValue([name]) });
		const { data, error } = await this.query(sql);
		if (error) return {
			data,
			error
		};
		else if (data.length === 0) return {
			data: null,
			error: { message: `Cannot find an extension named ${name}` }
		};
		else return {
			data: data[0],
			error
		};
	}
	async create({ name, schema, version, cascade = false }) {
		const sql = `
CREATE EXTENSION ${ident(name)}
  ${schema === void 0 ? "" : `SCHEMA ${ident(schema)}`}
  ${version === void 0 ? "" : `VERSION ${literal(version)}`}
  ${cascade ? "CASCADE" : ""};`;
		const { error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return await this.retrieve({ name });
	}
	async update(name, { update = false, version, schema }) {
		let updateSql = "";
		if (update) updateSql = `ALTER EXTENSION ${ident(name)} UPDATE ${version === void 0 ? "" : `TO ${literal(version)}`};`;
		const schemaSql = schema === void 0 ? "" : `ALTER EXTENSION ${ident(name)} SET SCHEMA ${ident(schema)};`;
		const sql = `BEGIN; ${updateSql} ${schemaSql} COMMIT;`;
		const { error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return await this.retrieve({ name });
	}
	async remove(name, { cascade = false } = {}) {
		const { data: extension, error } = await this.retrieve({ name });
		if (error) return {
			data: null,
			error
		};
		const sql = `DROP EXTENSION ${ident(name)} ${cascade ? "CASCADE" : "RESTRICT"};`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return {
			data: extension,
			error: null
		};
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/foreign_tables.sql.ts
const FOREIGN_TABLES_SQL = (props) => `
SELECT
  c.oid :: int8 AS id,
  n.nspname AS schema,
  c.relname AS name,
  obj_description(c.oid) AS comment
FROM
  pg_class c
  JOIN pg_namespace n ON n.oid = c.relnamespace
WHERE
  ${props.schemaFilter ? `n.nspname ${props.schemaFilter} AND` : ""}
  ${props.idsFilter ? `c.oid ${props.idsFilter} AND` : ""}
  ${props.tableIdentifierFilter ? `(n.nspname || '.' || c.relname) ${props.tableIdentifierFilter} AND` : ""}
  c.relkind = 'f'
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaForeignTables.ts
var PostgresMetaForeignTables = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includedSchemas, excludedSchemas, limit, offset, includeColumns = true } = {}) {
		const sql = generateEnrichedForeignTablesSql({
			includeColumns,
			schemaFilter: filterByList(includedSchemas, excludedSchemas),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name, schema = "public" }) {
		if (id) {
			const sql = generateEnrichedForeignTablesSql({
				includeColumns: true,
				idsFilter: filterByValue([`${id}`])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a foreign table with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name) {
			const sql = generateEnrichedForeignTablesSql({
				includeColumns: true,
				tableIdentifierFilter: filterByValue([`${schema}.${name}`])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a foreign table named ${name} in schema ${schema}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on foreign table retrieve" }
		};
	}
};
const generateEnrichedForeignTablesSql = ({ includeColumns, schemaFilter, idsFilter, tableIdentifierFilter, limit, offset }) => `
with foreign_tables as (${FOREIGN_TABLES_SQL({
	schemaFilter,
	tableIdentifierFilter,
	limit,
	offset
})})
  ${includeColumns ? `, columns as (${COLUMNS_SQL({
	schemaFilter,
	tableIdentifierFilter,
	tableIdFilter: idsFilter
})})` : ""}
select
  *
  ${includeColumns ? `, ${coalesceRowsToArray("columns", "columns.table_id = foreign_tables.id")}` : ""}
from foreign_tables`;

//#endregion
//#region src/vendor/postgres-meta/lib/sql/functions.sql.ts
const FUNCTIONS_SQL = (props) => `
-- CTE with sane arg_modes, arg_names, and arg_types.
-- All three are always of the same length.
-- All three include all args, including OUT and TABLE args.
with functions as (
  select
    p.*,
    -- proargmodes is null when all arg modes are IN
    coalesce(
      p.proargmodes,
      array_fill('i'::text, array[cardinality(coalesce(p.proallargtypes, p.proargtypes))])
    ) as arg_modes,
    -- proargnames is null when all args are unnamed
    coalesce(
      p.proargnames,
      array_fill(''::text, array[cardinality(coalesce(p.proallargtypes, p.proargtypes))])
    ) as arg_names,
    -- proallargtypes is null when all arg modes are IN
    coalesce(p.proallargtypes, p.proargtypes) as arg_types,
    array_cat(
      array_fill(false, array[pronargs - pronargdefaults]),
      array_fill(true, array[pronargdefaults])) as arg_has_defaults
  from
    pg_proc as p
    ${props.schemaFilter ? `join pg_namespace n on p.pronamespace = n.oid` : ""}
  where
    ${props.schemaFilter ? `n.nspname ${props.schemaFilter} AND` : ""}
    ${props.idsFilter ? `p.oid ${props.idsFilter} AND` : ""}
    ${props.nameFilter ? `p.proname ${props.nameFilter} AND` : ""}
    ${props.args === void 0 ? "" : props.args.length > 0 ? `p.proargtypes::text = ${props.args.length ? `(
          SELECT STRING_AGG(type_oid::text, ' ') FROM (
            SELECT (
              split_args.arr[
                array_length(
                  split_args.arr,
                  1
                )
              ]::regtype::oid
            ) AS type_oid FROM (
              SELECT STRING_TO_ARRAY(
                UNNEST(
                  ARRAY[${props.args}]
                ),
                ' '
              ) AS arr
            ) AS split_args
          ) args
    )` : "''"} AND` : ""}
    p.prokind = 'f'
)
select
  f.oid::int8 as id,
  n.nspname as schema,
  f.proname as name,
  l.lanname as language,
  case
    when l.lanname = 'internal' then ''
    else f.prosrc
  end as definition,
  case
    when l.lanname = 'internal' then f.prosrc
    else pg_get_functiondef(f.oid)
  end as complete_statement,
  coalesce(f_args.args, '[]') as args,
  pg_get_function_arguments(f.oid) as argument_types,
  pg_get_function_identity_arguments(f.oid) as identity_argument_types,
  f.prorettype::int8 as return_type_id,
  pg_get_function_result(f.oid) as return_type,
  nullif(rt.typrelid::int8, 0) as return_type_relation_id,
  f.proretset as is_set_returning_function,
  case
    when f.proretset then nullif(f.prorows, 0)
    else null
  end as prorows,
  case
    when f.provolatile = 'i' then 'IMMUTABLE'
    when f.provolatile = 's' then 'STABLE'
    when f.provolatile = 'v' then 'VOLATILE'
  end as behavior,
  f.prosecdef as security_definer,
  f_config.config_params as config_params
from
  functions f
  left join pg_namespace n on f.pronamespace = n.oid
  left join pg_language l on f.prolang = l.oid
  left join pg_type rt on rt.oid = f.prorettype
  left join (
    select
      oid,
      jsonb_object_agg(param, value) filter (where param is not null) as config_params
    from
      (
        select
          oid,
          (string_to_array(unnest(proconfig), '='))[1] as param,
          (string_to_array(unnest(proconfig), '='))[2] as value
        from
          functions
      ) as t
    group by
      oid
  ) f_config on f_config.oid = f.oid
  left join (
    select
      oid,
      jsonb_agg(jsonb_build_object(
        'mode', t2.mode,
        'name', name,
        'type_id', type_id,
        'has_default', has_default
      )) as args
    from
      (
        select
          oid,
          unnest(arg_modes) as mode,
          unnest(arg_names) as name,
          unnest(arg_types)::int8 as type_id,
          unnest(arg_has_defaults) as has_default
        from
          functions
      ) as t1,
      lateral (
        select
          case
            when t1.mode = 'i' then 'in'
            when t1.mode = 'o' then 'out'
            when t1.mode = 'b' then 'inout'
            when t1.mode = 'v' then 'variadic'
            else 'table'
          end as mode
      ) as t2
    group by
      t1.oid
  ) f_args on f_args.oid = f.oid
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaFunctions.ts
var PostgresMetaFunctions = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset } = {}) {
		const sql = FUNCTIONS_SQL({
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name, schema = "public", args = [] }) {
		const schemaFilter = schema ? filterByList([schema], []) : void 0;
		if (id) {
			const sql = FUNCTIONS_SQL({ idsFilter: filterByValue([id]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a function with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name && schema && args) {
			const sql = FUNCTIONS_SQL({
				schemaFilter,
				nameFilter: filterByValue([name]),
				args: args.map(literal)
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find function "${schema}"."${name}"(${args.join(", ")})` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on function retrieve" }
		};
	}
	async create({ name, schema = "public", args = [], definition, return_type = "void", language = "sql", behavior = "VOLATILE", security_definer = false, config_params = {} }) {
		const sql = this.generateCreateFunctionSql({
			name,
			schema,
			args,
			definition,
			return_type,
			language,
			behavior,
			security_definer,
			config_params
		});
		const { error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return await this.retrieve({
			name,
			schema,
			args
		});
	}
	async update(id, { name, schema, definition }) {
		const { data: currentFunc, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const args = currentFunc.argument_types.split(", ");
		const identityArgs = currentFunc.identity_argument_types;
		const updateDefinitionSql = typeof definition === "string" ? this.generateCreateFunctionSql({
			...currentFunc,
			definition,
			args,
			config_params: currentFunc.config_params ?? {}
		}, { replace: true }) : "";
		const updateNameSql = name && name !== currentFunc.name ? `ALTER FUNCTION ${ident(currentFunc.schema)}.${ident(currentFunc.name)}(${identityArgs}) RENAME TO ${ident(name)};` : "";
		const updateSchemaSql = schema && schema !== currentFunc.schema ? `ALTER FUNCTION ${ident(currentFunc.schema)}.${ident(name || currentFunc.name)}(${identityArgs})  SET SCHEMA ${ident(schema)};` : "";
		const currentSchemaFilter = currentFunc.schema ? filterByList([currentFunc.schema], []) : void 0;
		const currentNameFilter = currentFunc.name ? filterByValue([currentFunc.name]) : void 0;
		const sql = `
      DO LANGUAGE plpgsql $$
      BEGIN
        IF ${typeof definition === "string" ? "TRUE" : "FALSE"} THEN
          ${updateDefinitionSql}

          IF (
            SELECT id
            FROM (${FUNCTIONS_SQL({
			schemaFilter: currentSchemaFilter,
			nameFilter: currentNameFilter
		})}) AS f
            WHERE f.schema = ${literal(currentFunc.schema)}
            AND f.name = ${literal(currentFunc.name)}
            AND f.identity_argument_types = ${literal(identityArgs)}
          ) != ${id} THEN
            RAISE EXCEPTION 'Cannot find function "${currentFunc.schema}"."${currentFunc.name}"(${identityArgs})';
          END IF;
        END IF;

        ${updateNameSql}

        ${updateSchemaSql}
      END;
      $$;
    `;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return await this.retrieve({ id });
	}
	async remove(id, { cascade = false } = {}) {
		const { data: func, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const sql = `DROP FUNCTION ${ident(func.schema)}.${ident(func.name)}
    (${func.identity_argument_types})
    ${cascade ? "CASCADE" : "RESTRICT"};`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return {
			data: func,
			error: null
		};
	}
	generateCreateFunctionSql({ name, schema, args, definition, return_type, language, behavior, security_definer, config_params }, { replace = false } = {}) {
		return `
      CREATE ${replace ? "OR REPLACE" : ""} FUNCTION ${ident(schema)}.${ident(name)}(${args?.join(", ") || ""})
      RETURNS ${return_type}
      AS ${literal(definition)}
      LANGUAGE ${language}
      ${behavior}
      CALLED ON NULL INPUT
      ${security_definer ? "SECURITY DEFINER" : "SECURITY INVOKER"}
      ${config_params ? Object.entries(config_params).map(([param, value]) => `SET ${param} ${value[0] === "FROM CURRENT" ? "FROM CURRENT" : "TO " + value}`).join("\n") : ""};
    `;
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/indexes.sql.ts
const INDEXES_SQL = (props) => `
SELECT
    idx.indexrelid::int8 AS id,
    idx.indrelid::int8 AS table_id,
    n.nspname AS schema,
    idx.indnatts AS number_of_attributes,
    idx.indnkeyatts AS number_of_key_attributes,
    idx.indisunique AS is_unique,
    idx.indisprimary AS is_primary,
    idx.indisexclusion AS is_exclusion,
    idx.indimmediate AS is_immediate,
    idx.indisclustered AS is_clustered,
    idx.indisvalid AS is_valid,
    idx.indcheckxmin AS check_xmin,
    idx.indisready AS is_ready,
    idx.indislive AS is_live,
    idx.indisreplident AS is_replica_identity,
    idx.indkey::smallint[] AS key_attributes,
    idx.indcollation::integer[] AS collation,
    idx.indclass::integer[] AS class,
    idx.indoption::smallint[] AS options,
    idx.indpred AS index_predicate,
    obj_description(idx.indexrelid, 'pg_class') AS comment,
    ix.indexdef as index_definition,
    am.amname AS access_method,
    jsonb_agg(
      jsonb_build_object(
        'attribute_number', a.attnum,
        'attribute_name', a.attname,
        'data_type', format_type(a.atttypid, a.atttypmod)
      )
      ORDER BY a.attnum
    ) AS index_attributes
  FROM
    pg_index idx
    JOIN pg_class c ON c.oid = idx.indexrelid
    JOIN pg_namespace n ON c.relnamespace = n.oid
    JOIN pg_am am ON c.relam = am.oid
    JOIN pg_attribute a ON a.attrelid = c.oid AND a.attnum = ANY(idx.indkey)
    JOIN pg_indexes ix ON c.relname = ix.indexname
  WHERE
    ${props.schemaFilter ? `n.nspname ${props.schemaFilter}` : "true"}
    ${props.idsFilter ? `AND idx.indexrelid ${props.idsFilter}` : ""}
  GROUP BY
    idx.indexrelid, idx.indrelid, n.nspname, idx.indnatts, idx.indnkeyatts, idx.indisunique, idx.indisprimary, idx.indisexclusion, idx.indimmediate, idx.indisclustered, idx.indisvalid, idx.indcheckxmin, idx.indisready, idx.indislive, idx.indisreplident, idx.indkey, idx.indcollation, idx.indclass, idx.indoption, idx.indexprs, idx.indpred, ix.indexdef, am.amname
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaIndexes.ts
var PostgresMetaIndexes = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset } = {}) {
		const sql = INDEXES_SQL({
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id }) {
		if (id) {
			const sql = INDEXES_SQL({ idsFilter: filterByValue([id]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a index with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on function retrieve" }
		};
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/materialized_views.sql.ts
const MATERIALIZED_VIEWS_SQL = (props) => `
select
  c.oid::int8 as id,
  n.nspname as schema,
  c.relname as name,
  c.relispopulated as is_populated,
  obj_description(c.oid) as comment
from
  pg_class c
  join pg_namespace n on n.oid = c.relnamespace
where
  ${props.schemaFilter ? `n.nspname ${props.schemaFilter} AND` : ""}
  ${props.idsFilter ? `c.oid ${props.idsFilter} AND` : ""}
  ${props.materializedViewIdentifierFilter ? `(n.nspname || '.' || c.relname) ${props.materializedViewIdentifierFilter} AND` : ""}
  c.relkind = 'm'
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaMaterializedViews.ts
var PostgresMetaMaterializedViews = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includedSchemas, excludedSchemas, limit, offset, includeColumns = false } = {}) {
		let sql = generateEnrichedMaterializedViewsSql({
			includeColumns,
			schemaFilter: filterByList(includedSchemas, excludedSchemas, void 0),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name, schema = "public" }) {
		if (id) {
			const sql = generateEnrichedMaterializedViewsSql({
				includeColumns: true,
				idsFilter: filterByValue([id])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a materialized view with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name) {
			const sql = generateEnrichedMaterializedViewsSql({
				includeColumns: true,
				materializedViewIdentifierFilter: filterByValue([`${schema}.${name}`])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a materialized view named ${name} in schema ${schema}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on materialized view retrieve" }
		};
	}
};
const generateEnrichedMaterializedViewsSql = ({ includeColumns, schemaFilter, materializedViewIdentifierFilter, idsFilter, limit, offset }) => `
with materialized_views as (${MATERIALIZED_VIEWS_SQL({
	schemaFilter,
	limit,
	offset,
	materializedViewIdentifierFilter,
	idsFilter
})})
  ${includeColumns ? `, columns as (${COLUMNS_SQL({
	schemaFilter,
	limit,
	offset,
	tableIdentifierFilter: materializedViewIdentifierFilter,
	tableIdFilter: idsFilter
})})` : ""}
select
  *
  ${includeColumns ? `, ${coalesceRowsToArray("columns", "columns.table_id = materialized_views.id")}` : ""}
from materialized_views`;

//#endregion
//#region src/vendor/postgres-meta/lib/sql/policies.sql.ts
const POLICIES_SQL = (props) => `
SELECT
  pol.oid :: int8 AS id,
  n.nspname AS schema,
  c.relname AS table,
  c.oid :: int8 AS table_id,
  pol.polname AS name,
  CASE
    WHEN pol.polpermissive THEN 'PERMISSIVE' :: text
    ELSE 'RESTRICTIVE' :: text
  END AS action,
  CASE
    WHEN pol.polroles = '{0}' :: oid [] THEN array_to_json(
      string_to_array('public' :: text, '' :: text) :: name []
    )
    ELSE array_to_json(
      ARRAY(
        SELECT
          pg_roles.rolname
        FROM
          pg_roles
        WHERE
          pg_roles.oid = ANY (pol.polroles)
        ORDER BY
          pg_roles.rolname
      )
    )
  END AS roles,
  CASE
    pol.polcmd
    WHEN 'r' :: "char" THEN 'SELECT' :: text
    WHEN 'a' :: "char" THEN 'INSERT' :: text
    WHEN 'w' :: "char" THEN 'UPDATE' :: text
    WHEN 'd' :: "char" THEN 'DELETE' :: text
    WHEN '*' :: "char" THEN 'ALL' :: text
    ELSE NULL :: text
  END AS command,
  pg_get_expr(pol.polqual, pol.polrelid) AS definition,
  pg_get_expr(pol.polwithcheck, pol.polrelid) AS check
FROM
  pg_policy pol
  JOIN pg_class c ON c.oid = pol.polrelid
  LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
WHERE
  ${props.schemaFilter ? `n.nspname ${props.schemaFilter}` : "true"}
  ${props.idsFilter ? `AND pol.oid ${props.idsFilter}` : ""}
  ${props.functionNameIdentifierFilter ? `AND (c.relname || '.' || pol.polname) ${props.functionNameIdentifierFilter}` : ""}
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaPolicies.ts
var PostgresMetaPolicies = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset } = {}) {
		let sql = POLICIES_SQL({
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name, table, schema = "public" }) {
		const schemaFilter = schema ? filterByList([schema], []) : void 0;
		if (id) {
			const sql = POLICIES_SQL({ idsFilter: filterByValue([`${id}`]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a policy with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name && table) {
			const sql = POLICIES_SQL({
				schemaFilter,
				functionNameIdentifierFilter: filterByValue([`${table}.${name}`])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a policy named ${name} for table ${schema}.${table}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on policy retrieve" }
		};
	}
	async create({ name, table, schema = "public", definition, check, action = "PERMISSIVE", command = "ALL", roles = ["public"] }) {
		const definitionClause = definition === void 0 ? "" : `USING (${definition})`;
		const checkClause = check === void 0 ? "" : `WITH CHECK (${check})`;
		const sql = `
CREATE POLICY ${ident(name)} ON ${ident(schema)}.${ident(table)}
  AS ${action}
  FOR ${command}
  TO ${roles.map(ident).join(",")}
  ${definitionClause} ${checkClause};`;
		const { error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return await this.retrieve({
			name,
			table,
			schema
		});
	}
	async update(id, { name, definition, check, roles }) {
		const { data: old, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const alter = `ALTER POLICY ${ident(old.name)} ON ${ident(old.schema)}.${ident(old.table)}`;
		const nameSql = name === void 0 ? "" : `${alter} RENAME TO ${ident(name)};`;
		const sql = `BEGIN; ${definition === void 0 ? "" : `${alter} USING (${definition});`} ${check === void 0 ? "" : `${alter} WITH CHECK (${check});`} ${roles === void 0 ? "" : `${alter} TO ${roles.map(ident).join(",")};`} ${nameSql} COMMIT;`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return await this.retrieve({ id });
	}
	async remove(id) {
		const { data: policy, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const sql = `DROP POLICY ${ident(policy.name)} ON ${ident(policy.schema)}.${ident(policy.table)};`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return {
			data: policy,
			error: null
		};
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/publications.sql.ts
const PUBLICATIONS_SQL = (props) => `
SELECT
  p.oid :: int8 AS id,
  p.pubname AS name,
  p.pubowner::regrole::text AS owner,
  p.pubinsert AS publish_insert,
  p.pubupdate AS publish_update,
  p.pubdelete AS publish_delete,
  p.pubtruncate AS publish_truncate,
  CASE
    WHEN p.puballtables THEN NULL
    ELSE pr.tables
  END AS tables
FROM
  pg_catalog.pg_publication AS p
  LEFT JOIN LATERAL (
    SELECT
      COALESCE(
        array_agg(
          json_build_object(
            'id',
            c.oid :: int8,
            'name',
            c.relname,
            'schema',
            nc.nspname
          )
        ),
        '{}'
      ) AS tables
    FROM
      pg_catalog.pg_publication_rel AS pr
      JOIN pg_class AS c ON pr.prrelid = c.oid
      join pg_namespace as nc on c.relnamespace = nc.oid
    WHERE
      pr.prpubid = p.oid
  ) AS pr ON 1 = 1
WHERE
  ${props.idsFilter ? `p.oid ${props.idsFilter}` : "true"}
  ${props.nameFilter ? `AND p.pubname ${props.nameFilter}` : ""}
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaPublications.ts
var PostgresMetaPublications = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ limit, offset }) {
		let sql = PUBLICATIONS_SQL({
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name }) {
		if (id) {
			const sql = PUBLICATIONS_SQL({ idsFilter: filterByValue([id]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a publication with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name) {
			const sql = PUBLICATIONS_SQL({ nameFilter: filterByValue([name]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a publication named ${name}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on publication retrieve" }
		};
	}
	async create({ name, publish_insert = false, publish_update = false, publish_delete = false, publish_truncate = false, tables = null }) {
		let tableClause;
		if (tables === void 0 || tables === null) tableClause = "FOR ALL TABLES";
		else if (tables.length === 0) tableClause = "";
		else tableClause = `FOR TABLE ${tables.map((t) => {
			if (!t.includes(".")) return ident(t);
			const [schema, ...rest] = t.split(".");
			const table = rest.join(".");
			return `${ident(schema)}.${ident(table)}`;
		}).join(",")}`;
		let publishOps = [];
		if (publish_insert) publishOps.push("insert");
		if (publish_update) publishOps.push("update");
		if (publish_delete) publishOps.push("delete");
		if (publish_truncate) publishOps.push("truncate");
		const sql = `
CREATE PUBLICATION ${ident(name)} ${tableClause}
  WITH (publish = '${publishOps.join(",")}');`;
		const { error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return await this.retrieve({ name });
	}
	async update(id, { name, owner, publish_insert, publish_update, publish_delete, publish_truncate, tables }) {
		const sql = `
do $$
declare
  id oid := ${literal(id)};
  old record;
  new_name text := ${name === void 0 ? null : literal(name)};
  new_owner text := ${owner === void 0 ? null : literal(owner)};
  new_publish_insert bool := ${publish_insert ?? null};
  new_publish_update bool := ${publish_update ?? null};
  new_publish_delete bool := ${publish_delete ?? null};
  new_publish_truncate bool := ${publish_truncate ?? null};
  new_tables text := ${tables === void 0 ? null : literal(tables === null ? "all tables" : tables.map((t) => {
			if (!t.includes(".")) return ident(t);
			const [schema, ...rest] = t.split(".");
			const table = rest.join(".");
			return `${ident(schema)}.${ident(table)}`;
		}).join(","))};
begin
  select * into old from pg_publication where oid = id;
  if old is null then
    raise exception 'Cannot find publication with id %', id;
  end if;

  if new_tables is null then
    null;
  elsif new_tables = 'all tables' then
    if old.puballtables then
      null;
    else
      -- Need to recreate because going from list of tables <-> all tables with alter is not possible.
      execute(format('drop publication %1$I; create publication %1$I for all tables;', old.pubname));
    end if;
  else
    if old.puballtables then
      -- Need to recreate because going from list of tables <-> all tables with alter is not possible.
      execute(format('drop publication %1$I; create publication %1$I;', old.pubname));
    elsif exists(select from pg_publication_rel where prpubid = id) then
      execute(
        format(
          'alter publication %I drop table %s',
          old.pubname,
          (select string_agg(prrelid::regclass::text, ', ') from pg_publication_rel where prpubid = id)
        )
      );
    end if;

    -- At this point the publication must have no tables.

    if new_tables != '' then
      execute(format('alter publication %I add table %s', old.pubname, new_tables));
    end if;
  end if;

  execute(
    format(
      'alter publication %I set (publish = %L);',
      old.pubname,
      concat_ws(
        ', ',
        case when coalesce(new_publish_insert, old.pubinsert) then 'insert' end,
        case when coalesce(new_publish_update, old.pubupdate) then 'update' end,
        case when coalesce(new_publish_delete, old.pubdelete) then 'delete' end,
        case when coalesce(new_publish_truncate, old.pubtruncate) then 'truncate' end
      )
    )
  );

  execute(format('alter publication %I owner to %I;', old.pubname, coalesce(new_owner, old.pubowner::regrole::name)));

  -- Using the same name in the rename clause gives an error, so only do it if the new name is different.
  if new_name is not null and new_name != old.pubname then
    execute(format('alter publication %I rename to %I;', old.pubname, coalesce(new_name, old.pubname)));
  end if;

  -- We need to retrieve the publication later, so we need a way to uniquely identify which publication this is.
  -- We can't rely on id because it gets changed if it got recreated.
  -- We use a temp table to store the unique name - DO blocks can't return a value.
  create temp table pg_meta_publication_tmp (name) on commit drop as values (coalesce(new_name, old.pubname));
end $$;

with publications as (${PUBLICATIONS_SQL({})}) select * from publications where name = (select name from pg_meta_publication_tmp);
`;
		const { data, error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return {
			data: data[0],
			error
		};
	}
	async remove(id) {
		const { data: publication, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const sql = `DROP PUBLICATION IF EXISTS ${ident(publication.name)};`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return {
			data: publication,
			error: null
		};
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/table_relationships.sql.ts
const TABLE_RELATIONSHIPS_SQL = (props) => `
-- Adapted from
-- https://github.com/PostgREST/postgrest/blob/f9f0f79fa914ac00c11fbf7f4c558e14821e67e2/src/PostgREST/SchemaCache.hs#L722
WITH
pks_uniques_cols AS (
  SELECT
    connamespace,
    conrelid,
    jsonb_agg(column_info.cols) as cols
  FROM pg_constraint
  JOIN lateral (
    SELECT array_agg(cols.attname order by cols.attnum) as cols
    FROM ( select unnest(conkey) as col) _
    JOIN pg_attribute cols on cols.attrelid = conrelid and cols.attnum = col
  ) column_info ON TRUE
  WHERE
    contype IN ('p', 'u') and
    connamespace::regnamespace::text <> 'pg_catalog'
    ${props.schemaFilter ? `and connamespace::regnamespace::text ${props.schemaFilter}` : ""}
  GROUP BY connamespace, conrelid
)
SELECT
  traint.conname AS foreign_key_name,
  ns1.nspname AS schema,
  tab.relname AS relation,
  column_info.cols AS columns,
  ns2.nspname AS referenced_schema,
  other.relname AS referenced_relation,
  column_info.refs AS referenced_columns,
  (column_info.cols IN (SELECT * FROM jsonb_array_elements(pks_uqs.cols))) AS is_one_to_one
FROM pg_constraint traint
JOIN LATERAL (
  SELECT
    jsonb_agg(cols.attname order by ord) AS cols,
    jsonb_agg(refs.attname order by ord) AS refs
  FROM unnest(traint.conkey, traint.confkey) WITH ORDINALITY AS _(col, ref, ord)
  JOIN pg_attribute cols ON cols.attrelid = traint.conrelid AND cols.attnum = col
  JOIN pg_attribute refs ON refs.attrelid = traint.confrelid AND refs.attnum = ref
  WHERE ${props.schemaFilter ? `traint.connamespace::regnamespace::text ${props.schemaFilter}` : "true"}
) AS column_info ON TRUE
JOIN pg_namespace ns1 ON ns1.oid = traint.connamespace
JOIN pg_class tab ON tab.oid = traint.conrelid
JOIN pg_class other ON other.oid = traint.confrelid
JOIN pg_namespace ns2 ON ns2.oid = other.relnamespace
LEFT JOIN pks_uniques_cols pks_uqs ON pks_uqs.connamespace = traint.connamespace AND pks_uqs.conrelid = traint.conrelid
WHERE traint.contype = 'f'
AND traint.conparentid = 0
${props.schemaFilter ? `and ns1.nspname ${props.schemaFilter}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/sql/views_key_dependencies.sql.ts
const VIEWS_KEY_DEPENDENCIES_SQL = (props) => `
-- Adapted from
-- https://github.com/PostgREST/postgrest/blob/f9f0f79fa914ac00c11fbf7f4c558e14821e67e2/src/PostgREST/SchemaCache.hs#L820
with recursive
pks_fks as (
  -- pk + fk referencing col
  select
    contype::text as contype,
    conname,
    array_length(conkey, 1) as ncol,
    conrelid as resorigtbl,
    col as resorigcol,
    ord
  from pg_constraint
  left join lateral unnest(conkey) with ordinality as _(col, ord) on true
  where contype IN ('p', 'f')
  union
  -- fk referenced col
  select
    concat(contype, '_ref') as contype,
    conname,
    array_length(confkey, 1) as ncol,
    confrelid,
    col,
    ord
  from pg_constraint
  left join lateral unnest(confkey) with ordinality as _(col, ord) on true
  where contype='f'
  ${props.schemaFilter ? `and connamespace::regnamespace::text ${props.schemaFilter}` : ""}
),
views as (
  select
    c.oid       as view_id,
    n.nspname   as view_schema,
    c.relname   as view_name,
    r.ev_action as view_definition
  from pg_class c
  join pg_namespace n on n.oid = c.relnamespace
  join pg_rewrite r on r.ev_class = c.oid
  where c.relkind in ('v', 'm') 
    ${props.schemaFilter ? `and n.nspname ${props.schemaFilter}` : ""}
),
transform_json as (
  select
    view_id, view_schema, view_name,
    -- the following formatting is without indentation on purpose
    -- to allow simple diffs, with less whitespace noise
    replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      regexp_replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
      replace(
        view_definition::text,
      -- This conversion to json is heavily optimized for performance.
      -- The general idea is to use as few regexp_replace() calls as possible.
      -- Simple replace() is a lot faster, so we jump through some hoops
      -- to be able to use regexp_replace() only once.
      -- This has been tested against a huge schema with 250+ different views.
      -- The unit tests do NOT reflect all possible inputs. Be careful when changing this!
      -- -----------------------------------------------
      -- pattern           | replacement         | flags
      -- -----------------------------------------------
      -- <> in pg_node_tree is the same as null in JSON, but due to very poor performance of json_typeof
      -- we need to make this an empty array here to prevent json_array_elements from throwing an error
      -- when the targetList is null.
      -- We'll need to put it first, to make the node protection below work for node lists that start with
      -- null: (<> ..., too. This is the case for coldefexprs, when the first column does not have a default value.
         '<>'              , '()'
      -- , is not part of the pg_node_tree format, but used in the regex.
      -- This removes all , that might be part of column names.
      ), ','               , ''
      -- The same applies for { and }, although those are used a lot in pg_node_tree.
      -- We remove the escaped ones, which might be part of column names again.
      ), E'\\\\{'            , ''
      ), E'\\\\}'            , ''
      -- The fields we need are formatted as json manually to protect them from the regex.
      ), ' :targetList '   , ',"targetList":'
      ), ' :resno '        , ',"resno":'
      ), ' :resorigtbl '   , ',"resorigtbl":'
      ), ' :resorigcol '   , ',"resorigcol":'
      -- Make the regex also match the node type, e.g. \`{QUERY ...\`, to remove it in one pass.
      ), '{'               , '{ :'
      -- Protect node lists, which start with \`({\` or \`((\` from the greedy regex.
      -- The extra \`{\` is removed again later.
      ), '(('              , '{(('
      ), '({'              , '{({'
      -- This regex removes all unused fields to avoid the need to format all of them correctly.
      -- This leads to a smaller json result as well.
      -- Removal stops at \`,\` for used fields (see above) and \`}\` for the end of the current node.
      -- Nesting can't be parsed correctly with a regex, so we stop at \`{\` as well and
      -- add an empty key for the followig node.
      ), ' :[^}{,]+'       , ',"":'              , 'g'
      -- For performance, the regex also added those empty keys when hitting a \`,\` or \`}\`.
      -- Those are removed next.
      ), ',"":}'           , '}'
      ), ',"":,'           , ','
      -- This reverses the "node list protection" from above.
      ), '{('              , '('
      -- Every key above has been added with a \`,\` so far. The first key in an object doesn't need it.
      ), '{,'              , '{'
      -- pg_node_tree has \`()\` around lists, but JSON uses \`[]\`
      ), '('               , '['
      ), ')'               , ']'
      -- pg_node_tree has \` \` between list items, but JSON uses \`,\`
      ), ' '             , ','
    )::json as view_definition
  from views
),
target_entries as(
  select
    view_id, view_schema, view_name,
    json_array_elements(view_definition->0->'targetList') as entry
  from transform_json
),
results as(
  select
    view_id, view_schema, view_name,
    (entry->>'resno')::int as view_column,
    (entry->>'resorigtbl')::oid as resorigtbl,
    (entry->>'resorigcol')::int as resorigcol
  from target_entries
),
-- CYCLE detection according to PG docs: https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-CYCLE
-- Can be replaced with CYCLE clause once PG v13 is EOL.
recursion(view_id, view_schema, view_name, view_column, resorigtbl, resorigcol, is_cycle, path) as(
  select
    r.*,
    false,
    ARRAY[resorigtbl]
  from results r
  where ${props.schemaFilter ? `view_schema ${props.schemaFilter}` : "true"}
  union all
  select
    view.view_id,
    view.view_schema,
    view.view_name,
    view.view_column,
    tab.resorigtbl,
    tab.resorigcol,
    tab.resorigtbl = ANY(path),
    path || tab.resorigtbl
  from recursion view
  join results tab on view.resorigtbl=tab.view_id and view.resorigcol=tab.view_column
  where not is_cycle
),
repeated_references as(
  select
    view_id,
    view_schema,
    view_name,
    resorigtbl,
    resorigcol,
    array_agg(attname) as view_columns
  from recursion
  join pg_attribute vcol on vcol.attrelid = view_id and vcol.attnum = view_column
  group by
    view_id,
    view_schema,
    view_name,
    resorigtbl,
    resorigcol
)
select
  sch.nspname as table_schema,
  tbl.relname as table_name,
  rep.view_schema,
  rep.view_name,
  pks_fks.conname as constraint_name,
  pks_fks.contype as constraint_type,
  jsonb_agg(
    jsonb_build_object('table_column', col.attname, 'view_columns', view_columns) order by pks_fks.ord
  ) as column_dependencies
from repeated_references rep
join pks_fks using (resorigtbl, resorigcol)
join pg_class tbl on tbl.oid = rep.resorigtbl
join pg_attribute col on col.attrelid = tbl.oid and col.attnum = rep.resorigcol
join pg_namespace sch on sch.oid = tbl.relnamespace
group by sch.nspname, tbl.relname,  rep.view_schema, rep.view_name, pks_fks.conname, pks_fks.contype, pks_fks.ncol
-- make sure we only return key for which all columns are referenced in the view - no partial PKs or FKs
having ncol = array_length(array_agg(row(col.attname, view_columns) order by pks_fks.ord), 1)
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaRelationships.ts
var PostgresMetaRelationships = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas } = {}) {
		const schemaFilter = filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0);
		let allTableM2oAndO2oRelationships;
		{
			const sql = TABLE_RELATIONSHIPS_SQL({ schemaFilter });
			const { data, error } = await this.query(sql);
			if (error) return {
				data: null,
				error
			};
			allTableM2oAndO2oRelationships = data;
		}
		let allViewM2oAndO2oRelationships;
		{
			const viewsKeyDependenciesSql = VIEWS_KEY_DEPENDENCIES_SQL({ schemaFilter });
			const { data: viewsKeyDependencies, error } = await this.query(viewsKeyDependenciesSql);
			if (error) return {
				data: null,
				error
			};
			allViewM2oAndO2oRelationships = allTableM2oAndO2oRelationships.flatMap((r) => {
				const expandKeyDepCols = (colDeps) => {
					const tableColumns = colDeps.map(({ table_column }) => table_column);
					const cartesianProduct = (allEntries) => {
						return allEntries.reduce((results, entries) => results.map((result) => entries.map((entry) => result.concat(entry))).reduce((subResults, result) => subResults.concat(result), []), [[]]);
					};
					return cartesianProduct(colDeps.map((cd) => cd.view_columns)).map((viewColumns) => ({
						tableColumns,
						viewColumns
					}));
				};
				const viewToTableKeyDeps = viewsKeyDependencies.filter((vkd) => vkd.table_schema === r.schema && vkd.table_name === r.relation && vkd.constraint_name === r.foreign_key_name && vkd.constraint_type === "f");
				const tableToViewKeyDeps = viewsKeyDependencies.filter((vkd) => vkd.table_schema === r.referenced_schema && vkd.table_name === r.referenced_relation && vkd.constraint_name === r.foreign_key_name && vkd.constraint_type === "f_ref");
				const viewToTableRelationships = viewToTableKeyDeps.flatMap((vtkd) => expandKeyDepCols(vtkd.column_dependencies).map(({ viewColumns }) => ({
					foreign_key_name: r.foreign_key_name,
					schema: vtkd.view_schema,
					relation: vtkd.view_name,
					columns: viewColumns,
					is_one_to_one: r.is_one_to_one,
					referenced_schema: r.referenced_schema,
					referenced_relation: r.referenced_relation,
					referenced_columns: r.referenced_columns
				})));
				const tableToViewRelationships = tableToViewKeyDeps.flatMap((tvkd) => expandKeyDepCols(tvkd.column_dependencies).map(({ viewColumns }) => ({
					foreign_key_name: r.foreign_key_name,
					schema: r.schema,
					relation: r.relation,
					columns: r.columns,
					is_one_to_one: r.is_one_to_one,
					referenced_schema: tvkd.view_schema,
					referenced_relation: tvkd.view_name,
					referenced_columns: viewColumns
				})));
				const viewToViewRelationships = viewToTableKeyDeps.flatMap((vtkd) => expandKeyDepCols(vtkd.column_dependencies).flatMap(({ viewColumns }) => tableToViewKeyDeps.flatMap((tvkd) => expandKeyDepCols(tvkd.column_dependencies).map(({ viewColumns: referencedViewColumns }) => ({
					foreign_key_name: r.foreign_key_name,
					schema: vtkd.view_schema,
					relation: vtkd.view_name,
					columns: viewColumns,
					is_one_to_one: r.is_one_to_one,
					referenced_schema: tvkd.view_schema,
					referenced_relation: tvkd.view_name,
					referenced_columns: referencedViewColumns
				})))));
				return [
					...viewToTableRelationships,
					...tableToViewRelationships,
					...viewToViewRelationships
				];
			});
		}
		return {
			data: allTableM2oAndO2oRelationships.concat(allViewM2oAndO2oRelationships),
			error: null
		};
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/roles.sql.ts
const ROLES_SQL = (props) => `
-- TODO: Consider using pg_authid vs. pg_roles for unencrypted password field
SELECT
  oid :: int8 AS id,
  rolname AS name,
  rolsuper AS is_superuser,
  rolcreatedb AS can_create_db,
  rolcreaterole AS can_create_role,
  rolinherit AS inherit_role,
  rolcanlogin AS can_login,
  rolreplication AS is_replication_role,
  rolbypassrls AS can_bypass_rls,
  (
    SELECT
      COUNT(*)
    FROM
      pg_stat_activity
    WHERE
      pg_roles.rolname = pg_stat_activity.usename
  ) AS active_connections,
  CASE WHEN rolconnlimit = -1 THEN current_setting('max_connections') :: int8
       ELSE rolconnlimit
  END AS connection_limit,
  rolpassword AS password,
  rolvaliduntil AS valid_until,
  rolconfig AS config
FROM
  pg_roles
WHERE
  ${props.idsFilter ? `oid ${props.idsFilter}` : "true"}
  -- All default/predefined roles start with pg_: https://www.postgresql.org/docs/15/predefined-roles.html
  -- The pg_ prefix is also reserved.
  ${!props.includeDefaultRoles ? `AND NOT pg_catalog.starts_with(rolname, 'pg_')` : ""}
  ${props.nameFilter ? `AND rolname ${props.nameFilter}` : ""}
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaRoles.ts
function changeRoleConfig2Object(config) {
	if (!config) return null;
	return config.reduce((acc, cur) => {
		const [key, value] = cur.split("=");
		acc[key] = value;
		return acc;
	}, {});
}
var PostgresMetaRoles = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeDefaultRoles = false, limit, offset } = {}) {
		const sql = ROLES_SQL({
			limit,
			offset,
			includeDefaultRoles
		});
		const result = await this.query(sql);
		if (result.data) result.data = result.data.map((role) => {
			role.config = changeRoleConfig2Object(role.config);
			return role;
		});
		return result;
	}
	async retrieve({ id, name }) {
		if (id) {
			const sql = ROLES_SQL({ idsFilter: filterByValue([id]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a role with ID ${id}` }
			};
			else {
				data[0].config = changeRoleConfig2Object(data[0].config);
				return {
					data: data[0],
					error
				};
			}
		} else if (name) {
			const sql = ROLES_SQL({ nameFilter: filterByValue([name]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a role named ${name}` }
			};
			else {
				data[0].config = changeRoleConfig2Object(data[0].config);
				return {
					data: data[0],
					error
				};
			}
		} else return {
			data: null,
			error: { message: "Invalid parameters on role retrieve" }
		};
	}
	async create({ name, is_superuser = false, can_create_db = false, can_create_role = false, inherit_role = true, can_login = false, is_replication_role = false, can_bypass_rls = false, connection_limit = -1, password, valid_until, member_of, members, admins, config }) {
		const isSuperuserClause = is_superuser ? "SUPERUSER" : "NOSUPERUSER";
		const canCreateDbClause = can_create_db ? "CREATEDB" : "NOCREATEDB";
		const canCreateRoleClause = can_create_role ? "CREATEROLE" : "NOCREATEROLE";
		const inheritRoleClause = inherit_role ? "INHERIT" : "NOINHERIT";
		const canLoginClause = can_login ? "LOGIN" : "NOLOGIN";
		const isReplicationRoleClause = is_replication_role ? "REPLICATION" : "NOREPLICATION";
		const canBypassRlsClause = can_bypass_rls ? "BYPASSRLS" : "NOBYPASSRLS";
		const connectionLimitClause = `CONNECTION LIMIT ${connection_limit}`;
		const passwordClause = password === void 0 ? "" : `PASSWORD ${literal(password)}`;
		const validUntilClause = valid_until === void 0 ? "" : `VALID UNTIL ${literal(valid_until)}`;
		const memberOfClause = member_of === void 0 ? "" : `IN ROLE ${member_of.join(",")}`;
		const membersClause = members === void 0 ? "" : `ROLE ${members.join(",")}`;
		const adminsClause = admins === void 0 ? "" : `ADMIN ${admins.join(",")}`;
		let configClause = "";
		if (config !== void 0) configClause = Object.keys(config).map((k) => {
			const v = config[k];
			if (!k || !v) return "";
			return `ALTER ROLE ${name} SET ${k} = ${v};`;
		}).join("\n");
		const sql = `
BEGIN;
CREATE ROLE ${ident(name)}
WITH
  ${isSuperuserClause}
  ${canCreateDbClause}
  ${canCreateRoleClause}
  ${inheritRoleClause}
  ${canLoginClause}
  ${isReplicationRoleClause}
  ${canBypassRlsClause}
  ${connectionLimitClause}
  ${passwordClause}
  ${validUntilClause}
  ${memberOfClause}
  ${membersClause}
  ${adminsClause};
${configClause ? configClause : ""}
COMMIT;`;
		const { error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return await this.retrieve({ name });
	}
	async update(id, { name, is_superuser, can_create_db, can_create_role, inherit_role, can_login, is_replication_role, can_bypass_rls, connection_limit, password, valid_until, config }) {
		const { data: old, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const nameSql = name === void 0 ? "" : `ALTER ROLE ${ident(old.name)} RENAME TO ${ident(name)};`;
		let isSuperuserClause = "";
		if (is_superuser !== void 0) isSuperuserClause = is_superuser ? "SUPERUSER" : "NOSUPERUSER";
		let canCreateDbClause = "";
		if (can_create_db !== void 0) canCreateDbClause = can_create_db ? "CREATEDB" : "NOCREATEDB";
		let canCreateRoleClause = "";
		if (can_create_role !== void 0) canCreateRoleClause = can_create_role ? "CREATEROLE" : "NOCREATEROLE";
		let inheritRoleClause = "";
		if (inherit_role !== void 0) inheritRoleClause = inherit_role ? "INHERIT" : "NOINHERIT";
		let canLoginClause = "";
		if (can_login !== void 0) canLoginClause = can_login ? "LOGIN" : "NOLOGIN";
		let isReplicationRoleClause = "";
		if (is_replication_role !== void 0) isReplicationRoleClause = is_replication_role ? "REPLICATION" : "NOREPLICATION";
		let canBypassRlsClause = "";
		if (can_bypass_rls !== void 0) canBypassRlsClause = can_bypass_rls ? "BYPASSRLS" : "NOBYPASSRLS";
		const connectionLimitClause = connection_limit === void 0 ? "" : `CONNECTION LIMIT ${connection_limit}`;
		const passwordClause = password === void 0 ? "" : `PASSWORD ${literal(password)}`;
		const validUntilClause = valid_until === void 0 ? "" : `VALID UNTIL ${literal(valid_until)}`;
		let configClause = "";
		if (config !== void 0) configClause = config.map((c) => {
			const { op, path: path$1, value } = c;
			const k = path$1;
			const v = value || null;
			if (!k) throw new Error(`Invalid config value ${value}`);
			switch (op) {
				case "add":
				case "replace": return `ALTER ROLE ${ident(old.name)} SET ${ident(k)} = ${literal(v)};`;
				case "remove": return `ALTER ROLE ${ident(old.name)} RESET ${ident(k)};`;
				default: throw new Error(`Invalid config op ${op}`);
			}
		}).filter(Boolean).join("");
		const sql = `
BEGIN;
  ALTER ROLE ${ident(old.name)}
    ${isSuperuserClause}
    ${canCreateDbClause}
    ${canCreateRoleClause}
    ${inheritRoleClause}
    ${canLoginClause}
    ${isReplicationRoleClause}
    ${canBypassRlsClause}
    ${connectionLimitClause}
    ${passwordClause}
    ${validUntilClause};
  ${configClause ? configClause : ""}
  ${nameSql}
COMMIT;`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return await this.retrieve({ id });
	}
	async remove(id) {
		const { data: role, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const sql = `DROP ROLE ${ident(role.name)};`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return {
			data: role,
			error: null
		};
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/schemas.sql.ts
const SCHEMAS_SQL = (props) => `
-- Adapted from information_schema.schemata
select
  n.oid::int8 as id,
  n.nspname as name,
  u.rolname as owner
from
  pg_namespace n,
  pg_roles u
where
  n.nspowner = u.oid
  ${props.idsFilter ? `and n.oid ${props.idsFilter}` : ""}
  ${props.nameFilter ? `and n.nspname ${props.nameFilter}` : ""}
  ${!props.includeSystemSchemas ? `and not pg_catalog.starts_with(n.nspname, 'pg_')` : ""}
  and (
    pg_has_role(n.nspowner, 'USAGE')
    or has_schema_privilege(n.oid, 'CREATE, USAGE')
  )
  and not pg_catalog.starts_with(n.nspname, 'pg_temp_')
  and not pg_catalog.starts_with(n.nspname, 'pg_toast_temp_')
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaSchemas.ts
var PostgresMetaSchemas = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includedSchemas, excludedSchemas, includeSystemSchemas = false, limit, offset } = {}) {
		const sql = SCHEMAS_SQL({
			limit,
			offset,
			includeSystemSchemas,
			nameFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0)
		});
		return await this.query(sql);
	}
	async retrieve({ id, name }) {
		if (id) {
			const sql = SCHEMAS_SQL({ idsFilter: filterByValue([id]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a schema with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name) {
			const sql = SCHEMAS_SQL({ nameFilter: filterByValue([name]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a schema named ${name}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on schema retrieve" }
		};
	}
	async create({ name, owner = "postgres" }) {
		const sql = `CREATE SCHEMA ${ident(name)} AUTHORIZATION ${ident(owner)};`;
		const { error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return await this.retrieve({ name });
	}
	async update(id, { name, owner }) {
		const { data: old, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const nameSql = name === void 0 ? "" : `ALTER SCHEMA ${ident(old.name)} RENAME TO ${ident(name)};`;
		const sql = `BEGIN; ${owner === void 0 ? "" : `ALTER SCHEMA ${ident(old.name)} OWNER TO ${ident(owner)};`} ${nameSql} COMMIT;`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return await this.retrieve({ id });
	}
	async remove(id, { cascade = false } = {}) {
		const { data: schema, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const sql = `DROP SCHEMA ${ident(schema.name)} ${cascade ? "CASCADE" : "RESTRICT"};`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return {
			data: schema,
			error: null
		};
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/table_privileges.sql.ts
const TABLE_PRIVILEGES_SQL = (props) => `
-- Despite the name \`table_privileges\`, this includes other kinds of relations:
-- views, matviews, etc. "Relation privileges" just doesn't roll off the tongue.
--
-- For each relation, get its relacl in a jsonb format,
-- e.g.
--
-- '{postgres=arwdDxt/postgres}'
--
-- becomes
--
-- [
--   {
--     "grantee": "postgres",
--     "grantor": "postgres",
--     "is_grantable": false,
--     "privilege_type": "INSERT"
--   },
--   ...
-- ]
select
  c.oid as relation_id,
  nc.nspname as schema,
  c.relname as name,
  case
    when c.relkind = 'r' then 'table'
    when c.relkind = 'v' then 'view'
    when c.relkind = 'm' then 'materialized_view'
    when c.relkind = 'f' then 'foreign_table'
    when c.relkind = 'p' then 'partitioned_table'
  end as kind,
  coalesce(
    jsonb_agg(
      jsonb_build_object(
        'grantor', grantor.rolname,
        'grantee', grantee.rolname,
        'privilege_type', _priv.privilege_type,
        'is_grantable', _priv.is_grantable
      )
    ) filter (where _priv is not null),
    '[]'
  ) as privileges
from pg_class c
join pg_namespace as nc
  on nc.oid = c.relnamespace
left join lateral (
  select grantor, grantee, privilege_type, is_grantable
  from aclexplode(coalesce(c.relacl, acldefault('r', c.relowner)))
) as _priv on true
left join pg_roles as grantor
  on grantor.oid = _priv.grantor
left join (
  select
    pg_roles.oid,
    pg_roles.rolname
  from pg_roles
  union all
  select
    (0)::oid as oid, 'PUBLIC'
) as grantee (oid, rolname)
  on grantee.oid = _priv.grantee
where c.relkind in ('r', 'v', 'm', 'f', 'p')
  ${props.schemaFilter ? `and nc.nspname ${props.schemaFilter}` : ""}
  ${props.idsFilter ? `and c.oid ${props.idsFilter}` : ""}
  ${props.nameIdentifierFilter ? `and (nc.nspname || '.' || c.relname) ${props.nameIdentifierFilter}` : ""}
  and not pg_is_other_temp_schema(c.relnamespace)
  and (
    pg_has_role(c.relowner, 'USAGE')
    or has_table_privilege(
      c.oid,
      'SELECT, INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, TRIGGER, MAINTAIN'
    )
    or has_any_column_privilege(c.oid, 'SELECT, INSERT, UPDATE, REFERENCES')
  )
group by
  c.oid,
  nc.nspname,
  c.relname,
  c.relkind
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaTablePrivileges.ts
var PostgresMetaTablePrivileges = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset } = {}) {
		const sql = TABLE_PRIVILEGES_SQL({
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name, schema = "public" }) {
		if (id) {
			const sql = TABLE_PRIVILEGES_SQL({ idsFilter: filterByValue([id]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a relation with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name) {
			const sql = TABLE_PRIVILEGES_SQL({ nameIdentifierFilter: filterByValue([`${schema}.${name}`]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a relation named ${name} in schema ${schema}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on retrieving table privileges" }
		};
	}
	async grant(grants) {
		let sql = `
do $$
begin
${grants.map(({ privilege_type, relation_id, grantee, is_grantable }) => `execute format('grant ${privilege_type} on table %s to ${grantee.toLowerCase() === "public" ? "public" : ident(grantee)} ${is_grantable ? "with grant option" : ""}', ${relation_id}::regclass);`).join("\n")}
end $$;
`;
		const { data, error } = await this.query(sql);
		if (error) return {
			data,
			error
		};
		sql = TABLE_PRIVILEGES_SQL({ idsFilter: filterByList([...new Set(grants.map(({ relation_id }) => relation_id))]) });
		return await this.query(sql);
	}
	async revoke(revokes) {
		let sql = `
do $$
begin
${revokes.map((revoke) => `execute format('revoke ${revoke.privilege_type} on table %s from ${revoke.grantee}', ${revoke.relation_id}::regclass);`).join("\n")}
end $$;
`;
		const { data, error } = await this.query(sql);
		if (error) return {
			data,
			error
		};
		sql = TABLE_PRIVILEGES_SQL({ idsFilter: filterByList([...new Set(revokes.map(({ relation_id }) => relation_id))]) });
		return await this.query(sql);
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/triggers.sql.ts
const TRIGGERS_SQL = (props) => `
SELECT
  pg_t.oid AS id,
  pg_t.tgrelid AS table_id,
  CASE
    WHEN pg_t.tgenabled = 'D' THEN 'DISABLED'
    WHEN pg_t.tgenabled = 'O' THEN 'ORIGIN'
    WHEN pg_t.tgenabled = 'R' THEN 'REPLICA'
    WHEN pg_t.tgenabled = 'A' THEN 'ALWAYS'
    END AS enabled_mode,
  (
    STRING_TO_ARRAY(
      ENCODE(pg_t.tgargs, 'escape'), '\\000'
    )
  )[:pg_t.tgnargs] AS function_args,
  is_t.trigger_name AS name,
  is_t.event_object_table AS table,
  is_t.event_object_schema AS schema,
  is_t.action_condition AS condition,
  is_t.action_orientation AS orientation,
  is_t.action_timing AS activation,
  ARRAY_AGG(is_t.event_manipulation)::text[] AS events,
  pg_p.proname AS function_name,
  pg_n.nspname AS function_schema
FROM
  pg_trigger AS pg_t
JOIN
  pg_class AS pg_c
ON pg_t.tgrelid = pg_c.oid
JOIN pg_namespace AS table_ns
ON pg_c.relnamespace = table_ns.oid
JOIN information_schema.triggers AS is_t
ON is_t.trigger_name = pg_t.tgname
AND pg_c.relname = is_t.event_object_table
AND pg_c.relnamespace = (quote_ident(is_t.event_object_schema))::regnamespace
JOIN pg_proc AS pg_p
ON pg_t.tgfoid = pg_p.oid
JOIN pg_namespace AS pg_n
ON pg_p.pronamespace = pg_n.oid
WHERE
  ${props.schemaFilter ? `table_ns.nspname ${props.schemaFilter}` : "true"}
  ${props.tableNameFilter ? `AND pg_c.relname ${props.tableNameFilter}` : ""}
  ${props.nameFilter ? `AND is_t.trigger_name ${props.nameFilter}` : ""}
  ${props.idsFilter ? `AND pg_t.oid ${props.idsFilter}` : ""}
GROUP BY
  pg_t.oid,
  pg_t.tgrelid,
  pg_t.tgenabled,
  pg_t.tgargs,
  pg_t.tgnargs,
  is_t.trigger_name,
  is_t.event_object_table,
  is_t.event_object_schema,
  is_t.action_condition,
  is_t.action_orientation,
  is_t.action_timing,
  pg_p.proname,
  pg_n.nspname
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaTriggers.ts
var PostgresMetaTriggers = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset } = {}) {
		let sql = TRIGGERS_SQL({
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name, schema = "public", table }) {
		const schemaFilter = schema ? filterByList([schema], []) : void 0;
		if (id) {
			const sql = TRIGGERS_SQL({ idsFilter: filterByValue([id]) });
			const { data, error } = await this.query(sql);
			if (error) return {
				data: null,
				error
			};
			const triggerRecord = data && data[0];
			if (triggerRecord) return {
				data: triggerRecord,
				error: null
			};
			return {
				data: null,
				error: { message: `Cannot find a trigger with ID ${id}` }
			};
		}
		if (name && schema && table) {
			const sql = TRIGGERS_SQL({
				schemaFilter,
				nameFilter: filterByValue([name]),
				tableNameFilter: filterByValue([table])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data: null,
				error
			};
			const triggerRecord = data && data[0];
			if (triggerRecord) return {
				data: triggerRecord,
				error: null
			};
			return {
				data: null,
				error: { message: `Cannot find a trigger with name ${name} on table "${schema}"."${table}"` }
			};
		}
		return {
			data: null,
			error: { message: "Invalid parameters on trigger retrieve" }
		};
	}
	/**
	* Creates trigger
	*
	* @param {Object} obj - An object.
	* @param {string} obj.name - Trigger name.
	* @param {string} obj.schema - Name of schema that trigger is for.
	* @param {string} obj.table - Unqualified table, view, or foreign table name that trigger is for.
	* @param {string} obj.function_schema - Name of schema that function is for.
	* @param {string} obj.function_name - Unqualified name of the function to execute.
	* @param {('BEFORE'|'AFTER'|'INSTEAD OF')} obj.activation - Determines when function is called
	* during event occurrence.
	* @param {Array<string>} obj.events - Event(s) that will fire the trigger. Array of the following options: 'INSERT' | 'UPDATE' | 'UPDATE
	* OF column_name1,column_name2' | 'DELETE' | 'TRUNCATE'.
	* @param {('ROW'|'STATEMENT')} obj.orientation - Trigger function for every row affected by event or
	* once per statement. Defaults to 'STATEMENT'.
	* @param {string} obj.condition - Boolean expression that will trigger function.
	* For example: 'old.* IS DISTINCT FROM new.*'
	* @param {Array<string>} obj.function_args - array of arguments to be passed to function when trigger is fired.
	* For example: ['arg1', 'arg2']
	*/
	async create({ name, schema = "public", table, function_schema = "public", function_name, function_args, activation, events, orientation, condition }) {
		const qualifiedTableName = `${ident(schema)}.${ident(table)}`;
		const qualifiedFunctionName = `${ident(function_schema)}.${ident(function_name)}`;
		const triggerEvents = events.join(" OR ");
		const triggerOrientation = orientation ? `FOR EACH ${orientation}` : "";
		const triggerCondition = condition ? `WHEN (${condition})` : "";
		const functionArgs = `${function_args?.map(literal).join(",") ?? ""}`;
		const sql = `CREATE TRIGGER ${ident(name)} ${activation} ${triggerEvents} ON ${qualifiedTableName} ${triggerOrientation} ${triggerCondition} EXECUTE FUNCTION ${qualifiedFunctionName}(${functionArgs});`;
		const { error } = await this.query(sql);
		if (error) return {
			data: null,
			error
		};
		return await this.retrieve({
			name,
			table,
			schema
		});
	}
	async update(id, { name, enabled_mode }) {
		const { data: old, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		let enabledModeSql = "";
		switch (enabled_mode) {
			case "ORIGIN":
				enabledModeSql = `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ENABLE TRIGGER ${ident(old.name)};`;
				break;
			case "DISABLED":
				enabledModeSql = `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} DISABLE TRIGGER ${ident(old.name)};`;
				break;
			case "REPLICA":
			case "ALWAYS":
				enabledModeSql = `ALTER TABLE ${ident(old.schema)}.${ident(old.table)} ENABLE ${enabled_mode} TRIGGER ${ident(old.name)};`;
				break;
			default: break;
		}
		const nameSql = name && name !== old.name ? `ALTER TRIGGER ${ident(old.name)} ON ${ident(old.schema)}.${ident(old.table)} RENAME TO ${ident(name)};` : "";
		const sql = `BEGIN; ${enabledModeSql}; ${nameSql}; COMMIT;`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return await this.retrieve({ id });
	}
	async remove(id, { cascade = false } = {}) {
		const { data: triggerRecord, error } = await this.retrieve({ id });
		if (error) return {
			data: null,
			error
		};
		const { name, schema, table } = triggerRecord;
		const sql = `DROP TRIGGER ${ident(name)} ON ${ident(schema)}.${ident(table)} ${cascade ? "CASCADE" : ""};`;
		{
			const { error: error$1 } = await this.query(sql);
			if (error$1) return {
				data: null,
				error: error$1
			};
		}
		return {
			data: triggerRecord,
			error: null
		};
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/types.sql.ts
const TYPES_SQL = (props) => `
select
  t.oid::int8 as id,
  t.typname as name,
  n.nspname as schema,
  format_type (t.oid, null) as format,
  coalesce(t_enums.enums, '[]') as enums,
  coalesce(t_attributes.attributes, '[]') as attributes,
  obj_description (t.oid, 'pg_type') as comment,
  nullif(t.typrelid::int8, 0) as type_relation_id
from
  pg_type t
  left join pg_namespace n on n.oid = t.typnamespace
  left join (
    select
      enumtypid,
      jsonb_agg(enumlabel order by enumsortorder) as enums
    from
      pg_enum
    group by
      enumtypid
  ) as t_enums on t_enums.enumtypid = t.oid
  left join (
    select
      oid,
      jsonb_agg(
        jsonb_build_object('name', a.attname, 'type_id', a.atttypid::int8)
        order by a.attnum asc
      ) as attributes
    from
      pg_class c
      join pg_attribute a on a.attrelid = c.oid
    where
      c.relkind = 'c' and not a.attisdropped
    group by
      c.oid
  ) as t_attributes on t_attributes.oid = t.typrelid
  where
      (
        t.typrelid = 0
        or (
          select
            c.relkind ${props.includeTableTypes ? `in ('c', 'r', 'v', 'm', 'p')` : `= 'c'`}
          from
            pg_class c
          where
            c.oid = t.typrelid
        )
      )
      ${!props.includeArrayTypes ? `and not exists (
                 select
                 from
                   pg_type el
                 where
                   el.oid = t.typelem
                   and el.typarray = t.oid
               )` : ""}
      ${props.schemaFilter ? `and n.nspname ${props.schemaFilter}` : ""}
      ${props.idsFilter ? `and t.oid ${props.idsFilter}` : ""}
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaTypes.ts
var PostgresMetaTypes = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeTableTypes = false, includeArrayTypes = false, includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset } = {}) {
		const sql = TYPES_SQL({
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			limit,
			offset,
			includeTableTypes,
			includeArrayTypes
		});
		return await this.query(sql);
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/version.sql.ts
const VERSION_SQL = () => `
SELECT
  version(),
  current_setting('server_version_num') :: int8 AS version_number,
  (
    SELECT
      COUNT(*) AS active_connections
    FROM
      pg_stat_activity
  ) AS active_connections,
  current_setting('max_connections') :: int8 AS max_connections
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaVersion.ts
var PostgresMetaVersion = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async retrieve() {
		const { data, error } = await this.query(VERSION_SQL());
		if (error) return {
			data,
			error
		};
		return {
			data: data[0],
			error
		};
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/sql/views.sql.ts
const VIEWS_SQL = (props) => `
SELECT
  c.oid :: int8 AS id,
  n.nspname AS schema,
  c.relname AS name,
  -- See definition of information_schema.views
  (pg_relation_is_updatable(c.oid, false) & 20) = 20 AS is_updatable,
  obj_description(c.oid) AS comment
FROM
  pg_class c
  JOIN pg_namespace n ON n.oid = c.relnamespace
WHERE
  ${props.schemaFilter ? `n.nspname ${props.schemaFilter} AND` : ""}
  ${props.idsFilter ? `c.oid ${props.idsFilter} AND` : ""}
  ${props.viewIdentifierFilter ? `(n.nspname || '.' || c.relname) ${props.viewIdentifierFilter} AND` : ""}
  c.relkind = 'v'
${props.limit ? `limit ${props.limit}` : ""}
${props.offset ? `offset ${props.offset}` : ""}
`;

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMetaViews.ts
var PostgresMetaViews = class {
	query;
	constructor(query) {
		this.query = query;
	}
	async list({ includeSystemSchemas = false, includedSchemas, excludedSchemas, limit, offset, includeColumns = true } = {}) {
		const sql = generateEnrichedViewsSql({
			includeColumns,
			schemaFilter: filterByList(includedSchemas, excludedSchemas, !includeSystemSchemas ? DEFAULT_SYSTEM_SCHEMAS : void 0),
			limit,
			offset
		});
		return await this.query(sql);
	}
	async retrieve({ id, name, schema = "public" }) {
		if (id) {
			const sql = generateEnrichedViewsSql({
				includeColumns: true,
				idsFilter: filterByValue([id])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a view with ID ${id}` }
			};
			else return {
				data: data[0],
				error
			};
		} else if (name) {
			const sql = generateEnrichedViewsSql({
				includeColumns: true,
				viewIdentifierFilter: filterByValue([`${schema}.${name}`])
			});
			const { data, error } = await this.query(sql);
			if (error) return {
				data,
				error
			};
			else if (data.length === 0) return {
				data: null,
				error: { message: `Cannot find a view named ${name} in schema ${schema}` }
			};
			else return {
				data: data[0],
				error
			};
		} else return {
			data: null,
			error: { message: "Invalid parameters on view retrieve" }
		};
	}
};
const generateEnrichedViewsSql = ({ includeColumns, schemaFilter, idsFilter, viewIdentifierFilter, limit, offset }) => `
with views as (${VIEWS_SQL({
	schemaFilter,
	limit,
	offset,
	viewIdentifierFilter,
	idsFilter
})})
  ${includeColumns ? `, columns as (${COLUMNS_SQL({
	schemaFilter,
	tableIdentifierFilter: viewIdentifierFilter,
	tableIdFilter: idsFilter
})})` : ""}
select
  *
  ${includeColumns ? `, ${coalesceRowsToArray("columns", "columns.table_id = views.id")}` : ""}
from views`;

//#endregion
//#region src/vendor/postgres-meta/lib/db.ts
pg.types.setTypeParser(pg.types.builtins.INT8, (x) => {
	const asNumber = Number(x);
	if (Number.isSafeInteger(asNumber)) return asNumber;
	else return x;
});
pg.types.setTypeParser(pg.types.builtins.DATE, (x) => x);
pg.types.setTypeParser(pg.types.builtins.INTERVAL, (x) => x);
pg.types.setTypeParser(pg.types.builtins.TIMESTAMP, (x) => x);
pg.types.setTypeParser(pg.types.builtins.TIMESTAMPTZ, (x) => x);
pg.types.setTypeParser(1115, parse);
pg.types.setTypeParser(1182, parse);
pg.types.setTypeParser(1185, parse);
pg.types.setTypeParser(600, (x) => x);
pg.types.setTypeParser(1017, (x) => x);
const poolerQueryHandleError = (pgpool, sql, parameters) => {
	return new Promise((resolve, reject) => {
		let rejected = false;
		const connectionErrorHandler = (err) => {
			if (!rejected) setTimeout(() => {
				rejected = true;
				return reject(err);
			});
		};
		pgpool.once("error", connectionErrorHandler);
		pgpool.query(sql, parameters).then((results) => {
			if (!rejected) return resolve(results);
		}).catch((err) => {
			if (!rejected) {
				rejected = true;
				return reject(err);
			}
		});
	});
};
const init = (config) => {
	if (config.connectionString) {
		const u = new URL(config.connectionString);
		const sslmode = u.searchParams.get("sslmode");
		u.searchParams.delete("sslmode");
		u.searchParams.delete("sslca");
		u.searchParams.delete("sslkey");
		u.searchParams.delete("sslcert");
		u.searchParams.delete("sslrootcert");
		config.connectionString = u.toString();
		if (sslmode === null) {} else if (sslmode === "disable") config.ssl = false;
		else {
			if (typeof config.ssl !== "object") config.ssl = {};
			config.ssl.rejectUnauthorized = sslmode === "verify-full";
		}
	}
	let pool = new pg.Pool(config);
	return {
		async query(sql, { statementQueryTimeout, parameters } = {}) {
			const statementTimeoutQueryPrefix = statementQueryTimeout ? `SET statement_timeout='${statementQueryTimeout}s'; SET idle_session_timeout='${statementQueryTimeout}s';` : "";
			const sqlWithStatementTimeout = `${statementTimeoutQueryPrefix}${sql}`;
			try {
				if (!pool) {
					const transientPool = new pg.Pool(config);
					let res$1 = await poolerQueryHandleError(transientPool, sqlWithStatementTimeout, parameters);
					if (Array.isArray(res$1)) res$1 = res$1.reverse().find((x) => x.rows.length !== 0) ?? { rows: [] };
					await transientPool.end();
					return {
						data: res$1.rows,
						error: null
					};
				}
				let res = await poolerQueryHandleError(pool, sqlWithStatementTimeout, parameters);
				if (Array.isArray(res)) res = res.reverse().find((x) => x.rows.length !== 0) ?? { rows: [] };
				return {
					data: res.rows,
					error: null
				};
			} catch (error) {
				if (error.constructor.name === "DatabaseError") {
					let formattedError = "";
					if (error.severity) formattedError += `${error.severity}:  `;
					if (error.code) formattedError += `${error.code}: `;
					if (error.message) formattedError += error.message;
					formattedError += "\n";
					if (error.position) {
						const position = Number(error.position) - 1 - statementTimeoutQueryPrefix.length;
						error.position = `${position + 1}`;
						let line = "";
						let lineNumber = 0;
						let lineOffset = 0;
						const lines = sql.split("\n");
						let currentOffset = 0;
						for (let i = 0; i < lines.length; i++) {
							if (currentOffset + lines[i].length > position) {
								line = lines[i];
								lineNumber = i + 1;
								lineOffset = position - currentOffset;
								break;
							}
							currentOffset += lines[i].length + 1;
						}
						formattedError += `LINE ${lineNumber}: ${line}\n${" ".repeat(5 + lineNumber.toString().length + 2 + lineOffset)}^\n`;
					}
					if (error.detail) formattedError += `DETAIL:  ${error.detail}\n`;
					if (error.hint) formattedError += `HINT:  ${error.hint}\n`;
					if (error.internalQuery) formattedError += `QUERY:  ${error.internalQuery}\n`;
					if (error.where) formattedError += `CONTEXT:  ${error.where}\n`;
					return {
						data: null,
						error: {
							...error,
							message: error.message,
							formattedError
						}
					};
				}
				try {
					return {
						data: null,
						error: {
							code: error.code,
							message: error.message
						}
					};
				} finally {
					try {
						await this.end.bind(this);
					} catch (_endError) {
						console.error("Failed to end the connection on error: ", {
							this: this,
							end: this.end
						});
					}
				}
			}
		},
		async end() {
			try {
				const _pool = pool;
				pool = null;
				if (_pool) await _pool.end();
			} catch (endError) {
				console.error("Failed ending connection pool", endError);
			}
		}
	};
};

//#endregion
//#region src/vendor/postgres-meta/lib/PostgresMeta.ts
var PostgresMeta = class {
	query;
	end;
	columnPrivileges;
	columns;
	config;
	extensions;
	foreignTables;
	functions;
	indexes;
	materializedViews;
	policies;
	publications;
	relationships;
	roles;
	schemas;
	tablePrivileges;
	tables;
	triggers;
	types;
	version;
	views;
	constructor(config) {
		const { query, end } = init(config);
		this.query = query;
		this.end = end;
		this.columnPrivileges = new PostgresMetaColumnPrivileges(this.query);
		this.columns = new PostgresMetaColumns(this.query);
		this.config = new PostgresMetaConfig(this.query);
		this.extensions = new PostgresMetaExtensions(this.query);
		this.foreignTables = new PostgresMetaForeignTables(this.query);
		this.functions = new PostgresMetaFunctions(this.query);
		this.indexes = new PostgresMetaIndexes(this.query);
		this.materializedViews = new PostgresMetaMaterializedViews(this.query);
		this.policies = new PostgresMetaPolicies(this.query);
		this.publications = new PostgresMetaPublications(this.query);
		this.relationships = new PostgresMetaRelationships(this.query);
		this.roles = new PostgresMetaRoles(this.query);
		this.schemas = new PostgresMetaSchemas(this.query);
		this.tablePrivileges = new PostgresMetaTablePrivileges(this.query);
		this.tables = new PostgresMetaTables(this.query);
		this.triggers = new PostgresMetaTriggers(this.query);
		this.types = new PostgresMetaTypes(this.query);
		this.version = new PostgresMetaVersion(this.query);
		this.views = new PostgresMetaViews(this.query);
	}
};

//#endregion
//#region src/vendor/postgres-meta/lib/generators.ts
async function getGeneratorMetadata(pgMeta, filters = {
	includedSchemas: [],
	excludedSchemas: []
}) {
	const includedSchemas = filters.includedSchemas ?? [];
	const excludedSchemas = filters.excludedSchemas ?? [];
	const { data: schemas, error: schemasError } = await pgMeta.schemas.list({
		includeSystemSchemas: false,
		includedSchemas: includedSchemas.length > 0 ? includedSchemas : void 0,
		excludedSchemas: excludedSchemas.length > 0 ? excludedSchemas : void 0
	});
	if (schemasError) return {
		data: null,
		error: schemasError
	};
	const { data: tables, error: tablesError } = await pgMeta.tables.list({
		includedSchemas: includedSchemas.length > 0 ? includedSchemas : void 0,
		excludedSchemas: excludedSchemas.length > 0 ? excludedSchemas : void 0,
		includeColumns: false
	});
	if (tablesError) return {
		data: null,
		error: tablesError
	};
	const { data: foreignTables, error: foreignTablesError } = await pgMeta.foreignTables.list({
		includedSchemas: includedSchemas.length > 0 ? includedSchemas : void 0,
		excludedSchemas: excludedSchemas.length > 0 ? excludedSchemas : void 0,
		includeColumns: false
	});
	if (foreignTablesError) return {
		data: null,
		error: foreignTablesError
	};
	const { data: views, error: viewsError } = await pgMeta.views.list({
		includedSchemas: includedSchemas.length > 0 ? includedSchemas : void 0,
		excludedSchemas: excludedSchemas.length > 0 ? excludedSchemas : void 0,
		includeColumns: false
	});
	if (viewsError) return {
		data: null,
		error: viewsError
	};
	const { data: materializedViews, error: materializedViewsError } = await pgMeta.materializedViews.list({
		includedSchemas: includedSchemas.length > 0 ? includedSchemas : void 0,
		excludedSchemas: excludedSchemas.length > 0 ? excludedSchemas : void 0,
		includeColumns: false
	});
	if (materializedViewsError) return {
		data: null,
		error: materializedViewsError
	};
	const { data: columns, error: columnsError } = await pgMeta.columns.list({
		includedSchemas: includedSchemas.length > 0 ? includedSchemas : void 0,
		excludedSchemas: excludedSchemas.length > 0 ? excludedSchemas : void 0,
		includeSystemSchemas: false
	});
	if (columnsError) return {
		data: null,
		error: columnsError
	};
	const { data: relationships, error: relationshipsError } = await pgMeta.relationships.list({
		includedSchemas: includedSchemas.length > 0 ? includedSchemas : void 0,
		excludedSchemas: excludedSchemas.length > 0 ? excludedSchemas : void 0,
		includeSystemSchemas: false
	});
	if (relationshipsError) return {
		data: null,
		error: relationshipsError
	};
	const { data: functions, error: functionsError } = await pgMeta.functions.list({
		includedSchemas: includedSchemas.length > 0 ? includedSchemas : void 0,
		excludedSchemas: excludedSchemas.length > 0 ? excludedSchemas : void 0,
		includeSystemSchemas: false
	});
	if (functionsError) return {
		data: null,
		error: functionsError
	};
	const { data: types, error: typesError } = await pgMeta.types.list({
		includeTableTypes: true,
		includeArrayTypes: true,
		includeSystemSchemas: true
	});
	if (typesError) return {
		data: null,
		error: typesError
	};
	await pgMeta.end();
	return {
		data: {
			schemas: schemas.filter(({ name }) => !excludedSchemas.includes(name) && (includedSchemas.length === 0 || includedSchemas.includes(name))),
			tables,
			foreignTables,
			views,
			materializedViews,
			columns,
			relationships,
			functions: functions.filter(({ return_type }) => !["trigger", "event_trigger"].includes(return_type)),
			types
		},
		error: null
	};
}

//#endregion
//#region src/vendor/postgres-meta/constants.ts
const GENERATE_TYPES_DEFAULT_SCHEMA = process.env.PG_META_GENERATE_TYPES_DEFAULT_SCHEMA || "public";
const VALID_UNNAMED_FUNCTION_ARG_TYPES = new Set([
	114,
	3802,
	25
]);
const VALID_FUNCTION_ARGS_MODE = new Set([
	"in",
	"inout",
	"variadic"
]);

//#endregion
//#region src/vendor/postgres-meta/templates/typescript.ts
const apply = async ({ schemas, tables, foreignTables, views, materializedViews, columns, relationships, functions, types, detectOneToOneRelationships, postgrestVersion }) => {
	schemas.sort((a, b) => a.name.localeCompare(b.name));
	relationships.sort((a, b) => a.foreign_key_name.localeCompare(b.foreign_key_name) || a.referenced_relation.localeCompare(b.referenced_relation) || JSON.stringify(a.referenced_columns).localeCompare(JSON.stringify(b.referenced_columns)));
	const introspectionBySchema = Object.fromEntries(schemas.map((s) => [s.name, {
		tables: [],
		views: [],
		functions: [],
		enums: [],
		compositeTypes: []
	}]));
	const columnsByTableId = {};
	const tablesNamesByTableId = {};
	const relationTypeByIds = /* @__PURE__ */ new Map();
	const typesById = /* @__PURE__ */ new Map();
	const tablesLike = [
		...tables,
		...foreignTables,
		...views,
		...materializedViews
	];
	for (const tableLike of tablesLike) {
		columnsByTableId[tableLike.id] = [];
		tablesNamesByTableId[tableLike.id] = tableLike.name;
	}
	for (const column of columns) if (column.table_id in columnsByTableId) columnsByTableId[column.table_id].push(column);
	for (const tableId in columnsByTableId) columnsByTableId[tableId].sort((a, b) => a.name.localeCompare(b.name));
	for (const type of types) {
		typesById.set(type.id, type);
		if (type.type_relation_id) relationTypeByIds.set(type.id, type);
		if (type.schema in introspectionBySchema) {
			if (type.enums.length > 0) introspectionBySchema[type.schema].enums.push(type);
			if (type.attributes.length > 0) introspectionBySchema[type.schema].compositeTypes.push(type);
		}
	}
	function getRelationships(object, relationships$1) {
		return relationships$1.filter((relationship) => relationship.schema === object.schema && relationship.referenced_schema === object.schema && relationship.relation === object.name);
	}
	function generateRelationshiptTsDefinition(relationship) {
		return `{
      foreignKeyName: ${JSON.stringify(relationship.foreign_key_name)}
      columns: ${JSON.stringify(relationship.columns)}${detectOneToOneRelationships ? `\nisOneToOne: ${relationship.is_one_to_one}` : ""}
      referencedRelation: ${JSON.stringify(relationship.referenced_relation)}
      referencedColumns: ${JSON.stringify(relationship.referenced_columns)}
    }`;
	}
	for (const table of tables) if (table.schema in introspectionBySchema) introspectionBySchema[table.schema].tables.push({
		table,
		relationships: getRelationships(table, relationships)
	});
	for (const table of foreignTables) if (table.schema in introspectionBySchema) introspectionBySchema[table.schema].tables.push({
		table,
		relationships: getRelationships(table, relationships)
	});
	for (const view of views) if (view.schema in introspectionBySchema) introspectionBySchema[view.schema].views.push({
		view,
		relationships: getRelationships(view, relationships)
	});
	for (const materializedView of materializedViews) if (materializedView.schema in introspectionBySchema) introspectionBySchema[materializedView.schema].views.push({
		view: {
			...materializedView,
			is_updatable: false
		},
		relationships: getRelationships(materializedView, relationships)
	});
	const getTableNameFromRelationId = (relationId, returnTypeId) => {
		if (!relationId) return null;
		if (tablesNamesByTableId[relationId]) return tablesNamesByTableId[relationId];
		const reltype = returnTypeId ? relationTypeByIds.get(returnTypeId) : null;
		return reltype ? reltype.name : null;
	};
	for (const func of functions) if (func.schema in introspectionBySchema) {
		func.args.sort((a, b) => a.name.localeCompare(b.name));
		const inArgs = func.args.filter(({ mode }) => VALID_FUNCTION_ARGS_MODE.has(mode));
		if (inArgs.length === 0 || !inArgs.some(({ name }) => name === "") || inArgs.every((arg) => {
			if (arg.name === "") return arg.has_default && VALID_UNNAMED_FUNCTION_ARG_TYPES.has(arg.type_id);
			return true;
		}) || inArgs.length === 1 && inArgs[0].name === "" && (VALID_UNNAMED_FUNCTION_ARG_TYPES.has(inArgs[0].type_id) || relationTypeByIds.get(inArgs[0].type_id) && getTableNameFromRelationId(func.return_type_relation_id, func.return_type_id) || relationTypeByIds.get(inArgs[0].type_id) && !getTableNameFromRelationId(func.return_type_relation_id, func.return_type_id))) introspectionBySchema[func.schema].functions.push({
			fn: func,
			inArgs
		});
	}
	for (const schema in introspectionBySchema) {
		introspectionBySchema[schema].tables.sort((a, b) => a.table.name.localeCompare(b.table.name));
		introspectionBySchema[schema].views.sort((a, b) => a.view.name.localeCompare(b.view.name));
		introspectionBySchema[schema].functions.sort((a, b) => a.fn.name.localeCompare(b.fn.name));
		introspectionBySchema[schema].enums.sort((a, b) => a.name.localeCompare(b.name));
		introspectionBySchema[schema].compositeTypes.sort((a, b) => a.name.localeCompare(b.name));
	}
	const getFunctionTsReturnType = (fn, returnType) => {
		let setofOptionsInfo = "";
		const returnTableName = getTableNameFromRelationId(fn.return_type_relation_id, fn.return_type_id);
		const returnsSetOfTable = fn.is_set_returning_function && fn.return_type_relation_id !== null;
		const returnsMultipleRows = fn.prorows !== null && fn.prorows > 1;
		if (returnTableName) setofOptionsInfo = `SetofOptions: {
        from: "*"
        to: ${JSON.stringify(returnTableName)}
        isOneToOne: ${Boolean(!returnsMultipleRows)}
        isSetofReturn: ${fn.is_set_returning_function}
      }`;
		if (fn.args.length === 1) {
			const relationType = relationTypeByIds.get(fn.args[0].type_id);
			if (relationType) {
				const sourceTable = relationType.format;
				if (returnsSetOfTable && returnTableName) setofOptionsInfo = `SetofOptions: {
          from: ${JSON.stringify(sourceTable)}
          to: ${JSON.stringify(returnTableName)}
          isOneToOne: ${Boolean(!returnsMultipleRows)}
          isSetofReturn: true
        }`;
				else if (returnTableName && !returnsSetOfTable) {
					const targetTable = returnTableName;
					setofOptionsInfo = `SetofOptions: {
            from: ${JSON.stringify(sourceTable)}
            to: ${JSON.stringify(targetTable)}
            isOneToOne: true
            isSetofReturn: false
          }`;
				}
			}
		}
		return `${returnType}${fn.is_set_returning_function && returnsMultipleRows ? "[]" : ""}
                          ${setofOptionsInfo ? `${setofOptionsInfo}` : ""}`;
	};
	const getFunctionReturnType = (schema, fn) => {
		const tableArgs = fn.args.filter(({ mode }) => mode === "table");
		if (tableArgs.length > 0) return `{
              ${tableArgs.map(({ name, type_id }) => {
			const type$1 = typesById.get(type_id);
			let tsType = "unknown";
			if (type$1) tsType = pgTypeToTsType(schema, type$1.name, {
				types,
				schemas,
				tables,
				views
			});
			return {
				name,
				type: tsType
			};
		}).map(({ name, type: type$1 }) => `${JSON.stringify(name)}: ${type$1}`)}
            }`;
		const relation = introspectionBySchema[schema.name]?.tables.find(({ table: { id } }) => id === fn.return_type_relation_id)?.table || introspectionBySchema[schema.name]?.views.find(({ view: { id } }) => id === fn.return_type_relation_id)?.view;
		if (relation) return `{
              ${columnsByTableId[relation.id].map((column) => generateColumnTsDefinition(schema, {
			name: column.name,
			format: column.format,
			is_nullable: column.is_nullable,
			is_optional: false
		}, {
			types,
			schemas,
			tables,
			views
		})).join(",\n")}
            }`;
		const type = typesById.get(fn.return_type_id);
		if (type) return pgTypeToTsType(schema, type.name, {
			types,
			schemas,
			tables,
			views
		});
		return "unknown";
	};
	const hasTableRowError = (fn, inArgs) => {
		if (inArgs.length === 1 && inArgs[0].name === "" && relationTypeByIds.get(inArgs[0].type_id) && !getTableNameFromRelationId(fn.return_type_relation_id, fn.return_type_id)) return true;
		return false;
	};
	const getConflictError = (schema, fns, fn, inArgs) => {
		if (fns.length <= 1) return null;
		if (inArgs.length === 0) {
			const conflictingFns = fns.filter(({ fn: otherFn, inArgs: otherInArgs }) => {
				if (otherFn === fn) return false;
				return otherInArgs.length === 1 && otherInArgs[0].name === "" && otherInArgs[0].has_default;
			});
			if (conflictingFns.length > 0) {
				const conflictingFn = conflictingFns[0];
				const returnTypeName = typesById.get(conflictingFn.fn.return_type_id)?.name || "unknown";
				return `Could not choose the best candidate function between: ${schema.name}.${fn.name}(), ${schema.name}.${fn.name}( => ${returnTypeName}). Try renaming the parameters or the function itself in the database so function overloading can be resolved`;
			}
		}
		if (inArgs.length === 1 && inArgs[0].name !== "") {
			const conflictingFns = fns.filter(({ fn: otherFn, inArgs: otherInArgs }) => {
				if (otherFn === fn) return false;
				return otherInArgs.length === 1 && otherInArgs[0].name === inArgs[0].name && otherInArgs[0].type_id !== inArgs[0].type_id;
			});
			if (conflictingFns.length > 0) return `Could not choose the best candidate function between: ${[{
				fn,
				inArgs
			}, ...conflictingFns].sort((a, b) => {
				const aArgs = a.inArgs;
				const bArgs = b.inArgs;
				return (aArgs[0]?.type_id || 0) - (bArgs[0]?.type_id || 0);
			}).map((f) => {
				const args = f.inArgs;
				return `${schema.name}.${fn.name}(${args.map((a) => `${a.name || ""} => ${typesById.get(a.type_id)?.name || "unknown"}`).join(", ")})`;
			}).join(", ")}. Try renaming the parameters or the function itself in the database so function overloading can be resolved`;
		}
		return null;
	};
	const getFunctionSignatures = (schema, fns) => {
		return fns.map(({ fn, inArgs }) => {
			let argsType = "never";
			let returnType = getFunctionReturnType(schema, fn);
			const conflictError = getConflictError(schema, fns, fn, inArgs);
			if (conflictError) {
				if (inArgs.length > 0) argsType = `{ ${inArgs.map(({ name, type_id, has_default }) => {
					const type = typesById.get(type_id);
					let tsType = "unknown";
					if (type) tsType = pgTypeToTsType(schema, type.name, {
						types,
						schemas,
						tables,
						views
					});
					return {
						name,
						type: tsType,
						has_default
					};
				}).map(({ name, type, has_default }) => `${JSON.stringify(name)}${has_default ? "?" : ""}: ${type}`)} }`;
				returnType = `{ error: true } & ${JSON.stringify(conflictError)}`;
			} else if (hasTableRowError(fn, inArgs)) {
				if (inArgs.length > 0) argsType = `{ ${inArgs.map(({ name, type_id, has_default }) => {
					const type = typesById.get(type_id);
					let tsType = "unknown";
					if (type) tsType = pgTypeToTsType(schema, type.name, {
						types,
						schemas,
						tables,
						views
					});
					return {
						name,
						type: tsType,
						has_default
					};
				}).map(({ name, type, has_default }) => `${JSON.stringify(name)}${has_default ? "?" : ""}: ${type}`)} }`;
				returnType = `{ error: true } & ${JSON.stringify(`the function ${schema.name}.${fn.name} with parameter or with a single unnamed json/jsonb parameter, but no matches were found in the schema cache`)}`;
			} else if (inArgs.length > 0) argsType = `{ ${inArgs.map(({ name, type_id, has_default }) => {
				const type = typesById.get(type_id);
				let tsType = "unknown";
				if (type) tsType = pgTypeToTsType(schema, type.name, {
					types,
					schemas,
					tables,
					views
				});
				return {
					name,
					type: tsType,
					has_default
				};
			}).map(({ name, type, has_default }) => `${JSON.stringify(name)}${has_default ? "?" : ""}: ${type}`)} }`;
			return `{ Args: ${argsType}; Returns: ${getFunctionTsReturnType(fn, returnType)} }`;
		}).join(" |\n");
	};
	const internal_supabase_schema = postgrestVersion ? `// Allows to automatically instantiate createClient with right options
  // instead of createClient<Database, { PostgrestVersion: 'XX' }>(URL, KEY)
  __InternalSupabase: {
    PostgrestVersion: '${postgrestVersion}'
  }` : "";
	function generateNullableUnionTsType(tsType, isNullable) {
		if (tsType === "unknown" || tsType === "any" || !isNullable) return tsType;
		return `${tsType} | null`;
	}
	function generateColumnTsDefinition(schema, column, context) {
		return `${JSON.stringify(column.name)}${column.is_optional ? "?" : ""}: ${generateNullableUnionTsType(pgTypeToTsType(schema, column.format, context), column.is_nullable)}`;
	}
	let output = `
export type Json = string | number | boolean | null | { [key: string]: Json | undefined } | Json[]

export type Database = {
  ${internal_supabase_schema}
  ${schemas.map((schema) => {
		const { tables: schemaTables, views: schemaViews, functions: schemaFunctions, enums: schemaEnums, compositeTypes: schemaCompositeTypes } = introspectionBySchema[schema.name];
		return `${JSON.stringify(schema.name)}: {
          Tables: {
            ${schemaTables.length === 0 ? "[_ in never]: never" : schemaTables.map(({ table, relationships: relationships$1 }) => `${JSON.stringify(table.name)}: {
                  Row: {
                    ${[...columnsByTableId[table.id].map((column) => generateColumnTsDefinition(schema, {
			name: column.name,
			format: column.format,
			is_nullable: column.is_nullable,
			is_optional: false
		}, {
			types,
			schemas,
			tables,
			views
		})), ...schemaFunctions.filter(({ fn }) => fn.argument_types === table.name).map(({ fn }) => {
			return `${JSON.stringify(fn.name)}: ${generateNullableUnionTsType(getFunctionReturnType(schema, fn), true)}`;
		})]}
                  }
                  Insert: {
                    ${columnsByTableId[table.id].map((column) => {
			if (column.identity_generation === "ALWAYS") return `${JSON.stringify(column.name)}?: never`;
			return generateColumnTsDefinition(schema, {
				name: column.name,
				format: column.format,
				is_nullable: column.is_nullable,
				is_optional: column.is_nullable || column.is_identity || column.default_value !== null
			}, {
				types,
				schemas,
				tables,
				views
			});
		})}
                  }
                  Update: {
                    ${columnsByTableId[table.id].map((column) => {
			if (column.identity_generation === "ALWAYS") return `${JSON.stringify(column.name)}?: never`;
			return generateColumnTsDefinition(schema, {
				name: column.name,
				format: column.format,
				is_nullable: column.is_nullable,
				is_optional: true
			}, {
				types,
				schemas,
				tables,
				views
			});
		})}
                  }
                  Relationships: [
                    ${relationships$1.map(generateRelationshiptTsDefinition)}
                  ]
                }`)}
          }
          Views: {
            ${schemaViews.length === 0 ? "[_ in never]: never" : schemaViews.map(({ view, relationships: relationships$1 }) => `${JSON.stringify(view.name)}: {
                  Row: {
                    ${[...columnsByTableId[view.id].map((column) => generateColumnTsDefinition(schema, {
			name: column.name,
			format: column.format,
			is_nullable: column.is_nullable,
			is_optional: false
		}, {
			types,
			schemas,
			tables,
			views
		})), ...schemaFunctions.filter(({ fn }) => fn.argument_types === view.name).map(({ fn }) => `${JSON.stringify(fn.name)}: ${generateNullableUnionTsType(getFunctionReturnType(schema, fn), true)}`)]}
                  }
                  ${view.is_updatable ? `Insert: {
                           ${columnsByTableId[view.id].map((column) => {
			if (!column.is_updatable) return `${JSON.stringify(column.name)}?: never`;
			return generateColumnTsDefinition(schema, {
				name: column.name,
				format: column.format,
				is_nullable: true,
				is_optional: true
			}, {
				types,
				schemas,
				tables,
				views
			});
		})}
                         }
                         Update: {
                           ${columnsByTableId[view.id].map((column) => {
			if (!column.is_updatable) return `${JSON.stringify(column.name)}?: never`;
			return generateColumnTsDefinition(schema, {
				name: column.name,
				format: column.format,
				is_nullable: true,
				is_optional: true
			}, {
				types,
				schemas,
				tables,
				views
			});
		})}
                         }
                        ` : ""}Relationships: [
                    ${relationships$1.map(generateRelationshiptTsDefinition)}
                  ]
                }`)}
          }
          Functions: {
            ${(() => {
			if (schemaFunctions.length === 0) return "[_ in never]: never";
			const schemaFunctionsGroupedByName = schemaFunctions.reduce((acc, curr) => {
				acc[curr.fn.name] ??= [];
				acc[curr.fn.name].push(curr);
				return acc;
			}, {});
			for (const fnName in schemaFunctionsGroupedByName) schemaFunctionsGroupedByName[fnName].sort((a, b) => b.fn.definition.localeCompare(a.fn.definition));
			return Object.entries(schemaFunctionsGroupedByName).map(([fnName, fns]) => {
				const functionSignatures = getFunctionSignatures(schema, fns);
				return `${JSON.stringify(fnName)}:\n${functionSignatures}`;
			}).join(",\n");
		})()}
          }
          Enums: {
            ${schemaEnums.length === 0 ? "[_ in never]: never" : schemaEnums.map((enum_) => `${JSON.stringify(enum_.name)}: ${enum_.enums.map((variant) => JSON.stringify(variant)).join("|")}`)}
          }
          CompositeTypes: {
            ${schemaCompositeTypes.length === 0 ? "[_ in never]: never" : schemaCompositeTypes.map(({ name, attributes }) => `${JSON.stringify(name)}: {
                        ${attributes.map(({ name: name$1, type_id }) => {
			const type = typesById.get(type_id);
			let tsType = "unknown";
			if (type) tsType = `${generateNullableUnionTsType(pgTypeToTsType(schema, type.name, {
				types,
				schemas,
				tables,
				views
			}), true)}`;
			return `${JSON.stringify(name$1)}: ${tsType}`;
		})}
                      }`)}
          }
        }`;
	})}
}

type DatabaseWithoutInternals = Omit<Database, '__InternalSupabase'>

type DefaultSchema = DatabaseWithoutInternals[Extract<keyof Database, ${JSON.stringify(GENERATE_TYPES_DEFAULT_SCHEMA)}>]

export type Tables<
  DefaultSchemaTableNameOrOptions extends
    | keyof (DefaultSchema["Tables"] & DefaultSchema["Views"])
    | { schema: keyof DatabaseWithoutInternals },
  TableName extends DefaultSchemaTableNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals
  }
    ? keyof (DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"] &
        DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Views"])
    : never = never
> = DefaultSchemaTableNameOrOptions extends { schema: keyof DatabaseWithoutInternals }
  ? (DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"] &
      DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Views"])[TableName] extends {
      Row: infer R
    }
    ? R
    : never
  : DefaultSchemaTableNameOrOptions extends keyof (DefaultSchema["Tables"] & DefaultSchema["Views"])
  ? (DefaultSchema["Tables"] & DefaultSchema["Views"])[DefaultSchemaTableNameOrOptions] extends {
      Row: infer R
    }
    ? R
    : never
  : never

export type TablesInsert<
  DefaultSchemaTableNameOrOptions extends
    | keyof DefaultSchema["Tables"]
    | { schema: keyof DatabaseWithoutInternals },
  TableName extends DefaultSchemaTableNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals
  }
    ? keyof DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"]
    : never = never
> = DefaultSchemaTableNameOrOptions extends { schema: keyof DatabaseWithoutInternals }
  ? DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"][TableName] extends {
      Insert: infer I
    }
    ? I
    : never
  : DefaultSchemaTableNameOrOptions extends keyof DefaultSchema["Tables"]
  ? DefaultSchema["Tables"][DefaultSchemaTableNameOrOptions] extends {
      Insert: infer I
    }
    ? I
    : never
  : never

export type TablesUpdate<
  DefaultSchemaTableNameOrOptions extends
    | keyof DefaultSchema["Tables"]
    | { schema: keyof DatabaseWithoutInternals },
  TableName extends DefaultSchemaTableNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals
  }
    ? keyof DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"]
    : never = never
> = DefaultSchemaTableNameOrOptions extends { schema: keyof DatabaseWithoutInternals }
  ? DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"][TableName] extends {
      Update: infer U
    }
    ? U
    : never
  : DefaultSchemaTableNameOrOptions extends keyof DefaultSchema["Tables"]
  ? DefaultSchema["Tables"][DefaultSchemaTableNameOrOptions] extends {
      Update: infer U
    }
    ? U
    : never
  : never

export type Enums<
  DefaultSchemaEnumNameOrOptions extends
    | keyof DefaultSchema["Enums"]
    | { schema: keyof DatabaseWithoutInternals },
  EnumName extends DefaultSchemaEnumNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals
  }
    ? keyof DatabaseWithoutInternals[DefaultSchemaEnumNameOrOptions["schema"]]["Enums"]
    : never = never
> = DefaultSchemaEnumNameOrOptions extends { schema: keyof DatabaseWithoutInternals }
  ? DatabaseWithoutInternals[DefaultSchemaEnumNameOrOptions["schema"]]["Enums"][EnumName]
  : DefaultSchemaEnumNameOrOptions extends keyof DefaultSchema["Enums"]
  ? DefaultSchema["Enums"][DefaultSchemaEnumNameOrOptions]
  : never

export type CompositeTypes<
  PublicCompositeTypeNameOrOptions extends
    | keyof DefaultSchema["CompositeTypes"]
    | { schema: keyof DatabaseWithoutInternals },
  CompositeTypeName extends PublicCompositeTypeNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals
  }
    ? keyof DatabaseWithoutInternals[PublicCompositeTypeNameOrOptions["schema"]]["CompositeTypes"]
    : never = never
> = PublicCompositeTypeNameOrOptions extends { schema: keyof DatabaseWithoutInternals }
  ? DatabaseWithoutInternals[PublicCompositeTypeNameOrOptions["schema"]]["CompositeTypes"][CompositeTypeName]
  : PublicCompositeTypeNameOrOptions extends keyof DefaultSchema["CompositeTypes"]
  ? DefaultSchema["CompositeTypes"][PublicCompositeTypeNameOrOptions]
  : never

export const Constants = {
  ${schemas.map((schema) => {
		const schemaEnums = introspectionBySchema[schema.name].enums;
		return `${JSON.stringify(schema.name)}: {
          Enums: {
            ${schemaEnums.map((enum_) => `${JSON.stringify(enum_.name)}: [${enum_.enums.map((variant) => JSON.stringify(variant)).join(", ")}]`)}
          }
        }`;
	})}
} as const
`;
	output = await prettier.format(output, {
		parser: "typescript",
		semi: false
	});
	return output;
};
const pgTypeToTsType = (schema, pgType, { types, schemas, tables, views }) => {
	if (pgType === "bool") return "boolean";
	else if ([
		"int2",
		"int4",
		"int8",
		"float4",
		"float8",
		"numeric"
	].includes(pgType)) return "number";
	else if ([
		"bytea",
		"bpchar",
		"varchar",
		"date",
		"text",
		"citext",
		"time",
		"timetz",
		"timestamp",
		"timestamptz",
		"uuid",
		"vector"
	].includes(pgType)) return "string";
	else if (["json", "jsonb"].includes(pgType)) return "Json";
	else if (pgType === "void") return "undefined";
	else if (pgType === "record") return "Record<string, unknown>";
	else if (pgType.startsWith("_")) return `(${pgTypeToTsType(schema, pgType.substring(1), {
		types,
		schemas,
		tables,
		views
	})})[]`;
	else {
		const enumTypes = types.filter((type) => type.name === pgType && type.enums.length > 0);
		if (enumTypes.length > 0) {
			const enumType = enumTypes.find((type) => type.schema === schema.name) || enumTypes[0];
			if (schemas.some(({ name }) => name === enumType.schema)) return `Database[${JSON.stringify(enumType.schema)}]['Enums'][${JSON.stringify(enumType.name)}]`;
			return enumType.enums.map((variant) => JSON.stringify(variant)).join("|");
		}
		const compositeTypes = types.filter((type) => type.name === pgType && type.attributes.length > 0);
		if (compositeTypes.length > 0) {
			const compositeType = compositeTypes.find((type) => type.schema === schema.name) || compositeTypes[0];
			if (schemas.some(({ name }) => name === compositeType.schema)) return `Database[${JSON.stringify(compositeType.schema)}]['CompositeTypes'][${JSON.stringify(compositeType.name)}]`;
			return "unknown";
		}
		const tableRowTypes = tables.filter((table) => table.name === pgType);
		if (tableRowTypes.length > 0) {
			const tableRowType = tableRowTypes.find((type) => type.schema === schema.name) || tableRowTypes[0];
			if (schemas.some(({ name }) => name === tableRowType.schema)) return `Database[${JSON.stringify(tableRowType.schema)}]['Tables'][${JSON.stringify(tableRowType.name)}]['Row']`;
			return "unknown";
		}
		const viewRowTypes = views.filter((view) => view.name === pgType);
		if (viewRowTypes.length > 0) {
			const viewRowType = viewRowTypes.find((type) => type.schema === schema.name) || viewRowTypes[0];
			if (schemas.some(({ name }) => name === viewRowType.schema)) return `Database[${JSON.stringify(viewRowType.schema)}]['Views'][${JSON.stringify(viewRowType.name)}]['Row']`;
			return "unknown";
		}
		return "unknown";
	}
};

//#endregion
//#region src/cli/commands/generate-types.ts
async function generateTypes(options) {
	const { connectionString, schemas = ["public"], detectOneToOneRelationships = true } = options;
	const pgMeta = new PostgresMeta({ connectionString });
	try {
		const { data: metadata } = await getGeneratorMetadata(pgMeta, { includedSchemas: schemas });
		if (!metadata) throw new Error("No metadata found. Please check your database connection and schema.");
		return await apply({
			...metadata,
			detectOneToOneRelationships
		});
	} finally {
		await pgMeta.end();
	}
}

//#endregion
//#region src/cli/commands/gen-types.ts
async function genTypes(flags) {
	const dbUrl = flags.dbUrl?.trim();
	if (!dbUrl) {
		console.error("Error: --db-url is required");
		console.error("Run \"neon-js gen-types --help\" for usage information");
		process.exit(1);
	}
	const options = {
		connectionString: dbUrl,
		schemas: flags.schema,
		detectOneToOneRelationships: !flags.postgrestV9Compat
	};
	if (flags.queryTimeout) try {
		options.queryTimeout = parseDuration(flags.queryTimeout);
	} catch (error) {
		console.error(`Error: ${error instanceof Error ? error.message : error}`);
		process.exit(1);
	}
	const outputPath = flags.output;
	console.log("Generating types...");
	console.log(`Database: ${options.connectionString.split("@")[1] || "connected"}`);
	console.log(`Schemas: ${options.schemas?.join(", ") || "public"}`);
	console.log(`Output: ${outputPath}`);
	const types = await generateTypes(options);
	const outputDir = path.dirname(outputPath);
	if (outputDir !== ".") mkdirSync(outputDir, { recursive: true });
	writeFileSync(outputPath, types);
	console.log(`✅ Types generated successfully at ${outputPath}`);
}

//#endregion
//#region src/cli/index.ts
async function main() {
	const cli = meow(`
    Usage
      $ neon-js gen-types --db-url <url> [options]

    Commands
      gen-types    Generate TypeScript types from your database schema

    Required Options
      --db-url <url>              Database connection string

    Options
      --output, -o <path>         Output file path (default: database.types.ts)
      --schema, -s <name>         Schema to include (repeatable, default: public)
      --postgrest-v9-compat       Disable one-to-one relationship detection
      --query-timeout <duration>  Query timeout (default: 15s, format: 30s, 1m, 90s)

    Examples
      $ neon-js gen-types --db-url "postgresql://user:pass@host:5432/db"
      $ neon-js gen-types --db-url "postgresql://..." --output src/types/db.ts
      $ neon-js gen-types --db-url "postgresql://..." -s public -s auth
      $ neon-js gen-types --db-url "postgresql://..." --postgrest-v9-compat
      $ neon-js gen-types --db-url "postgresql://..." --query-timeout 30s
  `, {
		importMeta: import.meta,
		description: "A TypeScript SDK for Neon services",
		flags: {
			dbUrl: {
				type: "string",
				isRequired: (flags, input) => {
					if (flags.help || flags.version || input.length === 0) return false;
					return true;
				}
			},
			output: {
				type: "string",
				shortFlag: "o",
				default: "database.types.ts"
			},
			schema: {
				type: "string",
				shortFlag: "s",
				isMultiple: true,
				default: ["public"]
			},
			postgrestV9Compat: {
				type: "boolean",
				default: false
			},
			queryTimeout: { type: "string" }
		}
	});
	const command = cli.input[0];
	if (!command) {
		cli.showHelp(0);
		return;
	}
	if (cli.flags.help) {
		cli.showHelp(0);
		return;
	}
	if (command === "gen-types") await genTypes(cli.flags);
	else {
		console.error(`Unknown command: ${command}`);
		console.error("Run \"neon-js --help\" for usage information");
		process.exit(1);
	}
}
await main();

//#endregion
export {  };
